<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Preprocessing Full-Length Single-Cell RNA-Seq for Exon and Junction Read Counts &mdash; DOLPHIN  documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/copybutton.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/jquery.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
        <script src="../_static/clipboard.min.js"></script>
        <script src="../_static/copybutton.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Preprocessing 10X Single-Cell RNA-Seq for Exon and Junction Read Counts" href="step1_2_preprocess_10X.html" />
    <link rel="prev" title="Exon-Level Reference GTF File Generation" href="step0_generate_exon_gtf.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            DOLPHIN
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Getting started with DOLPHIN:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../usage.html">Usage</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="step0_generate_exon_gtf.html">Exon-Level Reference GTF File Generation</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Preprocessing Full-Length Single-Cell RNA-Seq for Exon and Junction Read Counts</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#step-1-download-required-tools">Step 1: Download Required Tools</a></li>
<li class="toctree-l3"><a class="reference internal" href="#step-2-create-a-reference-genome">Step 2: Create a Reference Genome</a></li>
<li class="toctree-l3"><a class="reference internal" href="#step-3-download-the-raw-rna-seq-files">Step 3: Download the Raw RNA-Seq Files</a></li>
<li class="toctree-l3"><a class="reference internal" href="#step4-trim">Step4: Trim</a></li>
<li class="toctree-l3"><a class="reference internal" href="#step5-star-alignment-align-to-modifed-exon-gtf-file-and-standard-reference-genome">Step5: STAR Alignment - align to modifed exon gtf file and standard reference genome</a></li>
<li class="toctree-l3"><a class="reference internal" href="#step-6-count-exon-reads-and-junction-reads">Step 6: Count Exon Reads and Junction Reads</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="step1_2_preprocess_10X.html">Preprocessing 10X Single-Cell RNA-Seq for Exon and Junction Read Counts</a></li>
<li class="toctree-l2"><a class="reference internal" href="step2_1_graph_generation.html">Generate Feature Matrix and Adjacency Matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="step2_2_exonGene_matrix_generation.html">Generate Exon Gene h5ad and Select Highly Variable Genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="step2_3_feature_matrix_generation.html">Generate Feature Matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="step2_4_adjacency_matrix_generation.html">Generate Adjacency Matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="step2_5_graph_data_generation.html">Generate Graph Data for MODEl input</a></li>
<li class="toctree-l2"><a class="reference internal" href="step3_run_DOLPHIN.html">Setup and Train the DOLPHIN Model</a></li>
<li class="toctree-l2"><a class="reference internal" href="step4_cell_aggregation.html">Cell Aggregation</a></li>
<li class="toctree-l2"><a class="reference internal" href="step5_alternative_splicing.html">Detect Alternative Splicing Events with Outrigger</a></li>
<li class="toctree-l2"><a class="reference internal" href="step6_alternative_splicing_analysis.html">Alternative Splicing Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="step6_alternative_splicing_analysis.html#Generate-PSI-adata">Generate PSI adata</a></li>
<li class="toctree-l2"><a class="reference internal" href="step6_alternative_splicing_analysis.html#Differential-Alternative-Splicing-Events">Differential Alternative Splicing Events</a></li>
<li class="toctree-l2"><a class="reference internal" href="step7_1_MAST.html">Run MAST to get Exon level p-value</a></li>
<li class="toctree-l2"><a class="reference internal" href="step7_2_EDEG_analysis.html">Exon-Level Differential Gene Analysis Using Feature Matrix</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Examples</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../examples/index.html">Examples</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API Documentations</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../API/index.html">API Documentations</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">DOLPHIN</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">Tutorials</a></li>
      <li class="breadcrumb-item active">Preprocessing Full-Length Single-Cell RNA-Seq for Exon and Junction Read Counts</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/tutorials/step1_1_preprocess_full_length.md.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="preprocessing-full-length-single-cell-rna-seq-for-exon-and-junction-read-counts">
<h1>Preprocessing Full-Length Single-Cell RNA-Seq for Exon and Junction Read Counts<a class="headerlink" href="#preprocessing-full-length-single-cell-rna-seq-for-exon-and-junction-read-counts" title="Permalink to this heading"></a></h1>
<p>Here is the brief pipeline for full-length and 10x single-cell RNA-seq shown:</p>
<img title="preprocess pipeline" alt="Alt text" src="preprocess_pipeline.png" />
<section id="step-1-download-required-tools">
<h2>Step 1: Download Required Tools<a class="headerlink" href="#step-1-download-required-tools" title="Permalink to this heading"></a></h2>
<p>Before starting the alignment process, make sure to download and install the following tools:</p>
<p><a class="reference external" href="https://github.com/alexdobin/STAR">STAR</a> &gt;=2.7.3a</p>
<p><a class="reference external" href="https://sourceforge.net/projects/subread/files/subread-2.0.8/">featurecounts</a> &gt;=2.0.3</p>
<p>optional:
<a class="reference external" href="http://www.usadellab.org/cms/index.php?page=trimmomatic">Trimmomatic</a> &gt;=0.39</p>
</section>
<section id="step-2-create-a-reference-genome">
<h2>Step 2: Create a Reference Genome<a class="headerlink" href="#step-2-create-a-reference-genome" title="Permalink to this heading"></a></h2>
<p>Run the following command to generate a reference genome for alignment using STAR.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">ensembl_mod_indx</span></code> is the directory where the reference genome index will be stored.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Homo_sapiens.GRCh38.dna_sm.primary_assembly.fa</span></code> can be downloaded <a class="reference external" href="https://ftp.ensembl.org/pub/release-113/fasta/homo_sapiens/dna/">here</a>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Homo_sapiens.GRCh38.107.exon.gtf</span></code> is generated using the <a class="reference internal" href="step0_generate_exon_gtf.html"><span class="std std-doc">file</span></a>.</p></li>
</ul>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>STAR<span class="w"> </span>--runMode<span class="w"> </span>genomeGenerate<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--genomeDir<span class="w"> </span>/mnt/data/kailu/STAR_example/ensembl_mod_indx/<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--genomeFastaFiles<span class="w"> </span>Homo_sapiens.GRCh38.dna_sm.primary_assembly.fa<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--sjdbGTFfile<span class="w"> </span>Homo_sapiens.GRCh38.107.exon.gtf<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--runThreadN<span class="w"> </span><span class="m">16</span>
</pre></div>
</div>
</section>
<section id="step-3-download-the-raw-rna-seq-files">
<h2>Step 3: Download the Raw RNA-Seq Files<a class="headerlink" href="#step-3-download-the-raw-rna-seq-files" title="Permalink to this heading"></a></h2>
<p>Download the raw RNA-seq files from the provided sources. For the links to the raw data, please refer to the original <a class="reference external" href="https://www.nature.com/articles/s41587-022-01312-3#data-availability">study</a>.</p>
<p>For full-length single-cell RNA-seq, each cell is stored in a separate FASTQ file. In the following steps, we will process one cell at a time. For example, the codes below processe cell ${ID_SAMPLE}</p>
</section>
<section id="step4-trim">
<h2>Step4: Trim<a class="headerlink" href="#step4-trim" title="Permalink to this heading"></a></h2>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># location of the timmomatic tools</span>
<span class="nv">trim</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;/mnt/data/kailu/Apps/Trimmomatic-0.39/trimmomatic-0.39.jar&quot;</span>

java<span class="w"> </span>-jar<span class="w"> </span><span class="nv">$trim</span><span class="w"> </span>SE<span class="w"> </span><span class="si">${</span><span class="nv">ID_SAMPLE</span><span class="si">}</span>.fastq.gz<span class="w"> </span><span class="si">${</span><span class="nv">ID_SAMPLE</span><span class="si">}</span>.trim.fastq.gz<span class="w"> </span>ILLUMINACLIP:/mnt/data/kailu/Apps/Trimmomatic-0.39/adapters/TruSeq3-SE.fa:2:30:10<span class="w"> </span>LEADING:3<span class="w"> </span>TRAILING:3<span class="w"> </span>SLIDINGWINDOW:4:15<span class="w"> </span>MINLEN:36<span class="w"> 	</span>
</pre></div>
</div>
</section>
<section id="step5-star-alignment-align-to-modifed-exon-gtf-file-and-standard-reference-genome">
<h2>Step5: STAR Alignment - align to modifed exon gtf file and standard reference genome<a class="headerlink" href="#step5-star-alignment-align-to-modifed-exon-gtf-file-and-standard-reference-genome" title="Permalink to this heading"></a></h2>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1">## `ID_SAMPLE` is the Cell Barcode Name</span>
mkdir<span class="w"> </span>./03_exon_star/<span class="si">${</span><span class="nv">ID_SAMPLE</span><span class="si">}</span>
STAR<span class="w"> </span>--runThreadN<span class="w"> </span><span class="m">4</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--genomeDir<span class="w"> </span>/mnt/data/kailu/STAR_example/ensembl_mod_indx/<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--readFilesIn<span class="w"> </span><span class="si">${</span><span class="nv">ID_SAMPLE</span><span class="si">}</span>.trim.fastq.gz<span class="w">  </span><span class="se">\</span>
<span class="w">    </span>--readFilesCommand<span class="w"> </span>gunzip<span class="w"> </span>-c<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--outSAMtype<span class="w"> </span>BAM<span class="w"> </span>SortedByCoordinate<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--outFileNamePrefix<span class="w"> </span>./03_exon_star/<span class="si">${</span><span class="nv">ID_SAMPLE</span><span class="si">}</span>/<span class="si">${</span><span class="nv">ID_SAMPLE</span><span class="si">}</span>.

mkdir<span class="w"> </span>./02_exon_std/<span class="si">${</span><span class="nv">ID_SAMPLE</span><span class="si">}</span>
STAR<span class="w"> </span>--runThreadN<span class="w"> </span><span class="m">4</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--genomeDir<span class="w"> </span>/mnt/data/kailu/STAR_example/ensembl_indx/<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--readFilesIn<span class="w"> </span><span class="si">${</span><span class="nv">ID_SAMPLE</span><span class="si">}</span>.trim.fastq.gz<span class="w">  </span><span class="se">\</span>
<span class="w">    </span>--readFilesCommand<span class="w"> </span>gunzip<span class="w"> </span>-c<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--outSAMtype<span class="w"> </span>BAM<span class="w"> </span>SortedByCoordinate<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--outFileNamePrefix<span class="w"> </span>./02_exon_std/<span class="si">${</span><span class="nv">ID_SAMPLE</span><span class="si">}</span>/<span class="si">${</span><span class="nv">ID_SAMPLE</span><span class="si">}</span>.std.
</pre></div>
</div>
</section>
<section id="step-6-count-exon-reads-and-junction-reads">
<h2>Step 6: Count Exon Reads and Junction Reads<a class="headerlink" href="#step-6-count-exon-reads-and-junction-reads" title="Permalink to this heading"></a></h2>
<p>Get exon gene count using the modified exon GTF file. This will generate the gene count (<code class="docutils literal notranslate"><span class="pre">${ID_SAMPLE}.exongene.count.txt</span></code>), which will be used later for HVG identification.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>mkdir<span class="w"> </span>./04_exon_gene_cnt
featureCounts<span class="w"> </span>-t<span class="w"> </span>exon<span class="w"> </span>-O<span class="w"> </span>-M<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-a<span class="w"> </span>Homo_sapiens.GRCh38.107.exon.gtf<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-o<span class="w"> </span>./04_exon_gene_cnt/<span class="si">${</span><span class="nv">ID_SAMPLE</span><span class="si">}</span>.exongene.count.txt<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>./03_exon_star/<span class="si">${</span><span class="nv">ID_SAMPLE</span><span class="si">}</span>/<span class="si">${</span><span class="nv">ID_SAMPLE</span><span class="si">}</span>.Aligned.sortedByCoord.out.bam
</pre></div>
</div>
<p>Run the following command to get the exon and junction counts. This step will generate the following files:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">${ID_SAMPLE}.exon.count.txt</span></code>: Exon read counts.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">${ID_SAMPLE}.exon.count.txt.jcounts</span></code>: Junction read counts.</p></li>
</ul>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>mkdir<span class="w"> </span>./05_exon_junct_cnt
featureCounts<span class="w"> </span>-t<span class="w"> </span>exon<span class="w"> </span>-f<span class="w"> </span>-O<span class="w"> </span>-J<span class="w"> </span>-M<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-a<span class="w"> </span>Homo_sapiens.GRCh38.107.exon.gtf<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-o<span class="w"> </span>./05_exon_junct_cnt/<span class="si">${</span><span class="nv">ID_SAMPLE</span><span class="si">}</span>.exon.count.txt<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>./03_exon_star/<span class="si">${</span><span class="nv">ID_SAMPLE</span><span class="si">}</span>/<span class="si">${</span><span class="nv">ID_SAMPLE</span><span class="si">}</span>.Aligned.sortedByCoord.out.bam
</pre></div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="step0_generate_exon_gtf.html" class="btn btn-neutral float-left" title="Exon-Level Reference GTF File Generation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="step1_2_preprocess_10X.html" class="btn btn-neutral float-right" title="Preprocessing 10X Single-Cell RNA-Seq for Exon and Junction Read Counts" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright Ding Lab at McGill University.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>