<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Exon-Level Differential Gene Analysis Using Feature Matrix &mdash; DOLPHIN  documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/copybutton.css" type="text/css" />
      <link rel="stylesheet" href="../_static/nbsphinx-code-cells.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/jquery.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
        <script src="../_static/clipboard.min.js"></script>
        <script src="../_static/copybutton.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Examples" href="../examples/index.html" />
    <link rel="prev" title="Run MAST to get Exon level p-value" href="step7_1_MAST.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            DOLPHIN
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Getting started with DOLPHIN:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../usage.html">Usage</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="step0_generate_exon_gtf.html">Exon-Level Reference GTF File Generation</a></li>
<li class="toctree-l2"><a class="reference internal" href="step1_1_preprocess_full_length.html">Preprocessing Full-Length Single-Cell RNA-Seq for Exon and Junction Read Counts</a></li>
<li class="toctree-l2"><a class="reference internal" href="step1_2_preprocess_10X.html">Preprocessing 10X Single-Cell RNA-Seq for Exon and Junction Read Counts</a></li>
<li class="toctree-l2"><a class="reference internal" href="step2_1_graph_generation.html">Generate Feature Matrix and Adjacency Matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="step2_2_exonGene_matrix_generation.html">Generate Exon Gene h5ad and Select Highly Variable Genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="step2_3_feature_matrix_generation.html">Generate Feature Matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="step2_4_adjacency_matrix_generation.html">Generate Adjacency Matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="step2_5_graph_data_generation.html">Generate Graph Data for MODEl input</a></li>
<li class="toctree-l2"><a class="reference internal" href="step3_run_DOLPHIN.html">Setup and Train the DOLPHIN Model</a></li>
<li class="toctree-l2"><a class="reference internal" href="step4_cell_aggregation.html">Cell Aggregation</a></li>
<li class="toctree-l2"><a class="reference internal" href="step5_alternative_splicing.html">Detect Alternative Splicing Events with Outrigger</a></li>
<li class="toctree-l2"><a class="reference internal" href="step6_alternative_splicing_analysis.html">Alternative Splicing Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="step6_alternative_splicing_analysis.html#Generate-PSI-adata">Generate PSI adata</a></li>
<li class="toctree-l2"><a class="reference internal" href="step6_alternative_splicing_analysis.html#Differential-Alternative-Splicing-Events">Differential Alternative Splicing Events</a></li>
<li class="toctree-l2"><a class="reference internal" href="step7_1_MAST.html">Run MAST to get Exon level p-value</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Exon-Level Differential Gene Analysis Using Feature Matrix</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#stouffer-method-and-average-p-value-and-average-log2fc">stouffer method and average p-value and average log2fc</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Examples</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../examples/index.html">Examples</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API Documentations</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../API/index.html">API Documentations</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">DOLPHIN</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">Tutorials</a></li>
      <li class="breadcrumb-item active">Exon-Level Differential Gene Analysis Using Feature Matrix</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/tutorials/step7_2_EDEG_analysis.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="Exon-Level-Differential-Gene-Analysis-Using-Feature-Matrix">
<h1>Exon-Level Differential Gene Analysis Using Feature Matrix<a class="headerlink" href="#Exon-Level-Differential-Gene-Analysis-Using-Feature-Matrix" title="Permalink to this heading">ÔÉÅ</a></h1>
<p>In this section, we perform exon-level differential gene analysis by utilizing the feature matrix. This analysis helps identify genes that show significant differences in expression between different conditions or cell types at the exon level, enabling a more granular understanding of gene regulation.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">statsmodels.stats.multitest</span> <span class="k">as</span> <span class="nn">smm</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">create_df_seurat</span><span class="p">(</span><span class="n">temp_name</span><span class="p">,</span> <span class="n">leiden_res</span><span class="p">,</span> <span class="n">num</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">temp_name</span><span class="p">)</span>
    <span class="n">pd_temp</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;./&quot;</span><span class="o">+</span><span class="n">temp_name</span><span class="o">+</span><span class="s2">&quot;/AdjacencyComp_&quot;</span><span class="o">+</span><span class="n">temp_name</span><span class="o">+</span><span class="s2">&quot;_&quot;</span><span class="o">+</span><span class="n">leiden_res</span><span class="o">+</span><span class="s2">&quot;_cluster_&quot;</span><span class="o">+</span><span class="n">num</span><span class="o">+</span><span class="s2">&quot;.csv&quot;</span><span class="p">)</span>
    <span class="n">pd_temp</span> <span class="o">=</span> <span class="n">pd_temp</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;Unnamed: 0&quot;</span><span class="p">:</span><span class="s2">&quot;Exon_names&quot;</span><span class="p">})</span>
    <span class="n">pd_temp</span><span class="p">[</span><span class="s1">&#39;Gene_names&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd_temp</span><span class="p">[</span><span class="s1">&#39;Exon_names&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[:</span><span class="n">x</span><span class="o">.</span><span class="n">rfind</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)]</span> <span class="k">if</span> <span class="s1">&#39;-&#39;</span> <span class="ow">in</span> <span class="n">x</span> <span class="k">else</span> <span class="n">x</span><span class="p">)</span>
    <span class="n">pd_temp</span> <span class="o">=</span> <span class="n">pd_temp</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;Gene_names&quot;</span><span class="p">:</span><span class="s2">&quot;Cancer_gene_names&quot;</span><span class="p">,</span> <span class="s2">&quot;Exon_names&quot;</span><span class="p">:</span><span class="s2">&quot;Cancer_names&quot;</span><span class="p">,</span><span class="s2">&quot;p_val&quot;</span><span class="p">:</span><span class="n">temp_name</span><span class="o">+</span><span class="s2">&quot;_pvals&quot;</span><span class="p">,</span> <span class="s2">&quot;p_val_adj&quot;</span><span class="p">:</span><span class="n">temp_name</span><span class="o">+</span><span class="s2">&quot;_pvals_adj&quot;</span><span class="p">,</span> <span class="s2">&quot;avg_log2FC&quot;</span><span class="p">:</span><span class="n">temp_name</span><span class="o">+</span><span class="s2">&quot;_logfoldchanges&quot;</span><span class="p">,</span> <span class="s2">&quot;pct.1&quot;</span><span class="p">:</span><span class="n">temp_name</span><span class="o">+</span><span class="s2">&quot;_pct.1&quot;</span><span class="p">,</span> <span class="s2">&quot;pct.2&quot;</span><span class="p">:</span><span class="n">temp_name</span><span class="o">+</span><span class="s2">&quot;_pct.2&quot;</span><span class="p">})</span>
    <span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">pd_temp</span><span class="p">[(</span><span class="n">pd_temp</span><span class="p">[</span><span class="n">temp_name</span><span class="o">+</span><span class="s2">&quot;_pvals_adj&quot;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mf">0.05</span><span class="p">)</span><span class="o">&amp;</span><span class="p">(</span><span class="nb">abs</span><span class="p">(</span><span class="n">pd_temp</span><span class="p">[</span><span class="n">temp_name</span><span class="o">+</span><span class="s2">&quot;_logfoldchanges&quot;</span><span class="p">])</span><span class="o">&gt;</span><span class="mi">1</span><span class="p">)][</span><span class="s2">&quot;Cancer_gene_names&quot;</span><span class="p">])))</span>
    <span class="k">return</span> <span class="n">pd_temp</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pd_combine</span> <span class="o">=</span> <span class="n">create_df_seurat</span><span class="p">(</span><span class="s2">&quot;MAST&quot;</span><span class="p">,</span> <span class="s2">&quot;leiden_0_4_8&quot;</span><span class="p">,</span> <span class="s2">&quot;2&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
MAST
234
</pre></div></div>
</div>
<section id="stouffer-method-and-average-p-value-and-average-log2fc">
<h2>stouffer method and average p-value and average log2fc<a class="headerlink" href="#stouffer-method-and-average-p-value-and-average-log2fc" title="Permalink to this heading">ÔÉÅ</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="nb">type</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;MAST&quot;</span><span class="p">]:</span>
    <span class="n">pd_combine</span><span class="p">[</span><span class="nb">type</span><span class="o">+</span><span class="s1">&#39;_abs_avg_log2FC&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd_combine</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;Cancer_gene_names&#39;</span><span class="p">)[</span><span class="nb">type</span><span class="o">+</span><span class="s1">&#39;_logfoldchanges&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span><span class="o">.</span><span class="n">abs</span><span class="p">()</span><span class="o">.</span><span class="n">mean</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">scipy.stats</span> <span class="kn">import</span> <span class="n">norm</span>

<span class="k">def</span> <span class="nf">stouffer_method</span><span class="p">(</span><span class="n">pvals</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Combine p-values using the unweighted Stouffer&#39;s method.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    - pvals: list or array of p-values</span>

<span class="sd">    Returns:</span>
<span class="sd">    - combined_p_value: The combined p-value after applying Stouffer&#39;s method.</span>
<span class="sd">    - combined_z: The combined z-score.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">non_nan_pvals</span> <span class="o">=</span> <span class="n">pvals</span><span class="p">[</span><span class="o">~</span><span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">pvals</span><span class="p">)]</span>

    <span class="k">if</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">non_nan_pvals</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>

    <span class="c1"># Convert p-values to z-scores</span>
    <span class="n">z_scores</span> <span class="o">=</span> <span class="n">norm</span><span class="o">.</span><span class="n">isf</span><span class="p">(</span><span class="n">non_nan_pvals</span><span class="p">)</span>  <span class="c1"># Inverse survival function (equivalent to 1 - CDF)</span>

    <span class="c1"># Calculate the unweighted z-score</span>
    <span class="n">combined_z</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">z_scores</span><span class="p">)</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">non_nan_pvals</span><span class="p">))</span>

    <span class="c1"># Convert the combined z-score back to a p-value</span>
    <span class="n">combined_p_value</span> <span class="o">=</span> <span class="n">norm</span><span class="o">.</span><span class="n">sf</span><span class="p">(</span><span class="n">combined_z</span><span class="p">)</span>  <span class="c1"># Survival function (1 - CDF)</span>

    <span class="k">return</span> <span class="n">combined_p_value</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="nb">type</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;MAST&quot;</span><span class="p">]:</span>
    <span class="n">temp_stouffer_p</span> <span class="o">=</span> <span class="n">pd_combine</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;Cancer_gene_names&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
        <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">stouffer_method</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="nb">type</span><span class="o">+</span><span class="s1">&#39;_pvals&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
    <span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;temp_p_value&#39;</span><span class="p">)</span>

    <span class="n">dict_temp_p_value</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">temp_stouffer_p</span><span class="p">[</span><span class="s1">&#39;Cancer_gene_names&#39;</span><span class="p">],</span> <span class="n">temp_stouffer_p</span><span class="p">[</span><span class="s1">&#39;temp_p_value&#39;</span><span class="p">]))</span>

    <span class="n">pd_combine</span><span class="p">[</span><span class="nb">type</span><span class="o">+</span><span class="s1">&#39;_stouffer_pval&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd_combine</span><span class="p">[</span><span class="s1">&#39;Cancer_gene_names&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">dict_temp_p_value</span><span class="p">)</span>

    <span class="c1">## adjusted weighted fisher&#39;s p-value</span>
    <span class="n">df_sub_temp</span> <span class="o">=</span> <span class="n">pd_combine</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="nb">type</span><span class="o">+</span><span class="s1">&#39;_stouffer_pval&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">df_sub_temp</span> <span class="o">=</span> <span class="n">df_sub_temp</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Cancer_gene_names&#39;</span><span class="p">])</span>
    <span class="n">_</span><span class="p">,</span> <span class="n">adjusted_pvals</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">smm</span><span class="o">.</span><span class="n">multipletests</span><span class="p">(</span><span class="n">df_sub_temp</span><span class="p">[</span><span class="nb">type</span><span class="o">+</span><span class="s1">&#39;_stouffer_pval&#39;</span><span class="p">],</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;fdr_bh&#39;</span><span class="p">)</span>
    <span class="n">_</span><span class="p">,</span> <span class="n">adjusted_pvals_bonf</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">smm</span><span class="o">.</span><span class="n">multipletests</span><span class="p">(</span><span class="n">df_sub_temp</span><span class="p">[</span><span class="nb">type</span><span class="o">+</span><span class="s1">&#39;_stouffer_pval&#39;</span><span class="p">],</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;bonferroni&#39;</span><span class="p">)</span>

    <span class="n">df_sub_temp</span><span class="p">[</span><span class="s2">&quot;temp_adj_p&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adjusted_pvals</span>
    <span class="n">df_sub_temp</span><span class="p">[</span><span class="s2">&quot;temp_adj_p_bonf&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adjusted_pvals_bonf</span>

    <span class="n">dict_p_adj</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">df_sub_temp</span><span class="p">[</span><span class="s2">&quot;Cancer_gene_names&quot;</span><span class="p">],</span> <span class="n">df_sub_temp</span><span class="p">[</span><span class="s2">&quot;temp_adj_p&quot;</span><span class="p">]))</span>
    <span class="n">dict_p_adj_bonf</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">df_sub_temp</span><span class="p">[</span><span class="s2">&quot;Cancer_gene_names&quot;</span><span class="p">],</span> <span class="n">df_sub_temp</span><span class="p">[</span><span class="s2">&quot;temp_adj_p_bonf&quot;</span><span class="p">]))</span>

    <span class="n">pd_combine</span><span class="p">[</span><span class="nb">type</span><span class="o">+</span><span class="s2">&quot;_stouffer_pval_adj&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd_combine</span><span class="p">[</span><span class="s2">&quot;Cancer_gene_names&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">dict_p_adj</span><span class="p">)</span>
    <span class="n">pd_combine</span><span class="p">[</span><span class="nb">type</span><span class="o">+</span><span class="s2">&quot;_stouffer_pval_adj_bonf&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd_combine</span><span class="p">[</span><span class="s2">&quot;Cancer_gene_names&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">dict_p_adj_bonf</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pd_combine_sub</span> <span class="o">=</span> <span class="n">pd_combine</span><span class="p">[(</span><span class="nb">abs</span><span class="p">(</span><span class="n">pd_combine</span><span class="p">[</span><span class="s2">&quot;MAST_abs_avg_log2FC&quot;</span><span class="p">])</span> <span class="o">&gt;</span><span class="mi">1</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">pd_combine</span><span class="p">[</span><span class="s2">&quot;MAST_stouffer_pval_adj_bonf&quot;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mf">0.05</span><span class="p">)]</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">pd_combine_sub</span><span class="p">[</span><span class="s2">&quot;Cancer_gene_names&quot;</span><span class="p">]))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
742
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pd_combine</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;./MAST/AdjacencyComp_leiden_0_4_8_cluster_2_MAST_v2_&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">pd_combine_sub</span><span class="p">[</span><span class="s2">&quot;Cancer_gene_names&quot;</span><span class="p">])))</span><span class="o">+</span><span class="s2">&quot;.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pd_combine_sub</span><span class="p">[</span><span class="s2">&quot;Cancer_gene_names&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;./MAST/AdjacencyComp_leiden_0_4_8_cluster_2_MAST_v2_Topp_&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">pd_combine_sub</span><span class="p">[</span><span class="s2">&quot;Cancer_gene_names&quot;</span><span class="p">])))</span><span class="o">+</span><span class="s2">&quot;.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="step7_1_MAST.html" class="btn btn-neutral float-left" title="Run MAST to get Exon level p-value" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../examples/index.html" class="btn btn-neutral float-right" title="Examples" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright Ding Lab at McGill University.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>