<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Generate Feature Matrix and Adjacency Matrix &mdash; DOLPHIN  documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/copybutton.css" type="text/css" />
      <link rel="stylesheet" href="../_static/nbsphinx-code-cells.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/jquery.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
        <script src="../_static/clipboard.min.js"></script>
        <script src="../_static/copybutton.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Generate Exon Gene h5ad and Select Highly Variable Genes" href="step2_2_exonGene_matrix_generation.html" />
    <link rel="prev" title="Preprocessing 10X Single-Cell RNA-Seq for Exon and Junction Read Counts" href="step1_2_preprocess_10X.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            DOLPHIN
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Getting started with DOLPHIN:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../usage.html">Usage</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="step0_generate_exon_gtf.html">Exon-Level Reference GTF File Generation</a></li>
<li class="toctree-l2"><a class="reference internal" href="step1_1_preprocess_full_length.html">Preprocessing Full-Length Single-Cell RNA-Seq for Exon and Junction Read Counts</a></li>
<li class="toctree-l2"><a class="reference internal" href="step1_2_preprocess_10X.html">Preprocessing 10X Single-Cell RNA-Seq for Exon and Junction Read Counts</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Generate Feature Matrix and Adjacency Matrix</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Metadata-File-Format">Metadata File Format</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Note-on-Processing">Note on Processing</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="step2_2_exonGene_matrix_generation.html">Generate Exon Gene h5ad and Select Highly Variable Genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="step2_3_feature_matrix_generation.html">Generate Feature Matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="step2_4_adjacency_matrix_generation.html">Generate Adjacency Matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="step2_5_graph_data_generation.html">Generate Graph Data for MODEl input</a></li>
<li class="toctree-l2"><a class="reference internal" href="step3_run_DOLPHIN.html">Setup and Train the DOLPHIN Model</a></li>
<li class="toctree-l2"><a class="reference internal" href="step4_cell_aggregation.html">Cell Aggregation</a></li>
<li class="toctree-l2"><a class="reference internal" href="step5_alternative_splicing.html">Detect Alternative Splicing Events with Outrigger</a></li>
<li class="toctree-l2"><a class="reference internal" href="step6_alternative_splicing_analysis.html">Alternative Splicing Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="step6_alternative_splicing_analysis.html#Generate-PSI-adata">Generate PSI adata</a></li>
<li class="toctree-l2"><a class="reference internal" href="step6_alternative_splicing_analysis.html#Differential-Alternative-Splicing-Events">Differential Alternative Splicing Events</a></li>
<li class="toctree-l2"><a class="reference internal" href="step7_1_MAST.html">Run MAST to get Exon level p-value</a></li>
<li class="toctree-l2"><a class="reference internal" href="step7_2_EDEG_analysis.html">Exon-Level Differential Gene Analysis Using Feature Matrix</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Examples</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../examples/index.html">Examples</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API Documentations</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../API/index.html">API Documentations</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">DOLPHIN</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">Tutorials</a></li>
      <li class="breadcrumb-item active">Generate Feature Matrix and Adjacency Matrix</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/tutorials/step2_1_graph_generation.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="Generate-Feature-Matrix-and-Adjacency-Matrix">
<h1>Generate Feature Matrix and Adjacency Matrix<a class="headerlink" href="#Generate-Feature-Matrix-and-Adjacency-Matrix" title="Permalink to this heading"></a></h1>
<p>Generate the <strong>feature matrix</strong> and <strong>adjacency matrix</strong> using the exon read counts and junction read counts obtained from the count files. For detailed instructions on aligning the RNA-seq data, refer to the following tutorials:</p>
<ul class="simple">
<li><p><strong>Full-length RNA-seq</strong>: <a class="reference internal" href="step1_1_preprocess_full_length.html"><span class="doc">step1_1_preprocess_full_length.md</span></a></p></li>
<li><p><strong>10X RNA-seq</strong>: <a class="reference internal" href="step1_2_preprocess_10X.html"><span class="doc">step1_2_preprocess_10X.md</span></a></p></li>
</ul>
<p>Before running this section, please ensure that your counts data and metadata file is ready, and also make sure you have the <code class="docutils literal notranslate"><span class="pre">adj_index.csv</span></code> and <code class="docutils literal notranslate"><span class="pre">gtf.pkl</span></code> files prepared. These files can be generated in this <a class="reference internal" href="step0_generate_exon_gtf.html"><span class="doc">file</span></a>.</p>
<p>For human GRCh38, you can directly download the necessary files from <a class="reference external" href="https://outlook.office.com/host/377c982d-9686-450e-9a7c-22aeaf1bc162/7211f19f-262a-42eb-a02e-289956491741">here</a>.</p>
<section id="Metadata-File-Format">
<h2>Metadata File Format<a class="headerlink" href="#Metadata-File-Format" title="Permalink to this heading"></a></h2>
<p>The metadata file should be tab-separated, with the <strong>“CB”</strong> (cell barcode) column as the only mandatory column. You can add as many metadata columns as needed. These additional metadata columns will be retained in all output <code class="docutils literal notranslate"><span class="pre">adata</span></code> files as <strong>obs</strong> (observations).</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>CB</p></th>
<th class="head"><p>cluster</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>T8_AAAGCAAGTCGCGAAA</p></td>
<td><p>Stellate cell</p></td>
</tr>
<tr class="row-odd"><td><p>T8_AAATGCCAGCTGGAAC</p></td>
<td><p>Macrophage cell</p></td>
</tr>
<tr class="row-even"><td><p>T8_AAATGCCGTAGCTGCC</p></td>
<td><p>Macrophage cell</p></td>
</tr>
<tr class="row-odd"><td><p>T8_AAATGCCTCCACTGGG</p></td>
<td><p>Ductal cell type 2</p></td>
</tr>
<tr class="row-even"><td><p>…</p></td>
<td><p>…</p></td>
</tr>
</tbody>
</table>
<p><strong>Notes:</strong></p>
<ul class="simple">
<li><p>The <strong>“CB”</strong> column is mandatory as it represents the cell barcode.</p></li>
<li><p>Additional columns (e.g., <strong>cluster</strong>, <strong>source</strong>, etc.) can be added based on the specific metadata required for your analysis.</p></li>
<li><p>These metadata columns will be included in the output <code class="docutils literal notranslate"><span class="pre">adata</span></code> files under <strong>obs</strong>.</p></li>
<li><p><strong>“CB”</strong> will be used to find the <strong>count table file names</strong> and will be stored as <strong>obs index</strong> in <code class="docutils literal notranslate"><span class="pre">adata</span></code>.</p></li>
</ul>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">DOLPHIN</span> <span class="kn">import</span> <span class="n">gene</span><span class="p">,</span> <span class="n">get_gtf</span>
<span class="kn">import</span> <span class="nn">os</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/mnt/md0/kailu/anaconda3/envs/DOLPHIN/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## This folder contains all the counts table files, undert this main folder &quot;05_exon_junct_cnt&quot; contains exon counts and junction counts files</span>
<span class="n">main_folder</span> <span class="o">=</span> <span class="s2">&quot;./00_data_generation/&quot;</span>
<span class="n">output_folder</span> <span class="o">=</span> <span class="s2">&quot;06_graph_mtx&quot;</span> <span class="c1">## output files are stored here</span>
<span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">main_folder</span><span class="p">,</span> <span class="n">output_folder</span><span class="p">),</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">metadata</span> <span class="o">=</span> <span class="s2">&quot;your_metaData.csv&quot;</span>
<span class="n">pd_gt</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">metadata</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">mr_cb_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">pd_gt</span><span class="p">[</span><span class="n">pd_gt</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">0</span><span class="p">]])</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">path_gtf</span> <span class="o">=</span> <span class="s2">&quot;./gtf.pkl&quot;</span>
<span class="n">path_adj_index</span> <span class="o">=</span> <span class="s2">&quot;./adj_index.csv&quot;</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gtf</span><span class="p">,</span> <span class="n">adj_ind</span> <span class="o">=</span> <span class="n">get_gtf</span><span class="p">(</span><span class="n">path_gtf</span><span class="p">,</span> <span class="n">path_adj_index</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Note-on-Processing">
<h2>Note on Processing<a class="headerlink" href="#Note-on-Processing" title="Permalink to this heading"></a></h2>
<p>This will run one cell at a time. Later, implementing multiprocessing will help speed up the process by running multiple cells concurrently.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">mr_cb_list</span><span class="p">)):</span>
    <span class="n">g</span> <span class="o">=</span> <span class="n">gene</span><span class="p">(</span><span class="n">main_folder</span><span class="p">,</span> <span class="n">gtf</span><span class="p">,</span> <span class="n">adj_ind</span><span class="p">,</span> <span class="n">mr_cb_list</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
    <span class="n">g</span><span class="o">.</span><span class="n">get_all</span><span class="p">()</span>
</pre></div>
</div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="step1_2_preprocess_10X.html" class="btn btn-neutral float-left" title="Preprocessing 10X Single-Cell RNA-Seq for Exon and Junction Read Counts" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="step2_2_exonGene_matrix_generation.html" class="btn btn-neutral float-right" title="Generate Exon Gene h5ad and Select Highly Variable Genes" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright Ding Lab at McGill University.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>