<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Setup and Train the DOLPHIN Model &mdash; DOLPHIN  documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/copybutton.css" type="text/css" />
      <link rel="stylesheet" href="../_static/nbsphinx-code-cells.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/jquery.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
        <script src="../_static/clipboard.min.js"></script>
        <script src="../_static/copybutton.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Cell Aggregation" href="step4_cell_aggregation.html" />
    <link rel="prev" title="Generate Graph Data for MODEl input" href="step2_5_graph_data_generation.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            DOLPHIN
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Getting started with DOLPHIN:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../usage.html">Usage</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="step0_generate_exon_gtf.html">Exon-Level Reference GTF File Generation</a></li>
<li class="toctree-l2"><a class="reference internal" href="step1_1_preprocess_full_length.html">Preprocessing Full-Length Single-Cell RNA-Seq for Exon and Junction Read Counts</a></li>
<li class="toctree-l2"><a class="reference internal" href="step1_2_preprocess_10X.html">Preprocessing 10X Single-Cell RNA-Seq for Exon and Junction Read Counts</a></li>
<li class="toctree-l2"><a class="reference internal" href="step2_1_graph_generation.html">Generate Feature Matrix and Adjacency Matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="step2_2_exonGene_matrix_generation.html">Generate Exon Gene h5ad and Select Highly Variable Genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="step2_3_feature_matrix_generation.html">Generate Feature Matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="step2_4_adjacency_matrix_generation.html">Generate Adjacency Matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="step2_5_graph_data_generation.html">Generate Graph Data for MODEl input</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Setup and Train the DOLPHIN Model</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Load-Processed-Dataset">Load Processed Dataset</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Set-Hyperparameters-and-Train-the-Model">Set Hyperparameters and Train the Model</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Function-Definition"><strong>Function Definition</strong></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#Cell-Embedding-Cluster-Using-X_z">Cell Embedding Cluster Using <code class="docutils literal notranslate"><span class="pre">X_z</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="step4_cell_aggregation.html">Cell Aggregation</a></li>
<li class="toctree-l2"><a class="reference internal" href="step5_alternative_splicing.html">Detect Alternative Splicing Events with Outrigger</a></li>
<li class="toctree-l2"><a class="reference internal" href="step6_alternative_splicing_analysis.html">Alternative Splicing Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="step6_alternative_splicing_analysis.html#Generate-PSI-adata">Generate PSI adata</a></li>
<li class="toctree-l2"><a class="reference internal" href="step6_alternative_splicing_analysis.html#Differential-Alternative-Splicing-Events">Differential Alternative Splicing Events</a></li>
<li class="toctree-l2"><a class="reference internal" href="step7_1_MAST.html">Run MAST to get Exon level p-value</a></li>
<li class="toctree-l2"><a class="reference internal" href="step7_2_EDEG_analysis.html">Exon-Level Differential Gene Analysis Using Feature Matrix</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Examples</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../examples/index.html">Examples</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API Documentations</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../API/index.html">API Documentations</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">DOLPHIN</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">Tutorials</a></li>
      <li class="breadcrumb-item active">Setup and Train the DOLPHIN Model</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/tutorials/step3_run_DOLPHIN.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="Setup-and-Train-the-DOLPHIN-Model">
<h1>Setup and Train the DOLPHIN Model<a class="headerlink" href="#Setup-and-Train-the-DOLPHIN-Model" title="Permalink to this heading"></a></h1>
<p>This tutorial provides a step-by-step guide on configuring the model architecture, setting hyperparameters, and visualizing cell embedding clusters using DOLPHIN.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">DOLPHIN</span> <span class="kn">import</span> <span class="n">run_DOLPHIN</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
</pre></div>
</div>
</div>
<section id="Load-Processed-Dataset">
<h2>Load Processed Dataset<a class="headerlink" href="#Load-Processed-Dataset" title="Permalink to this heading"></a></h2>
<p>Specify the graph data input and the highly variable gene (HVG)-filtered feature matrix obtained from the preprocessing step.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#load datasets</span>
<span class="n">graph_data</span> <span class="o">=</span> <span class="s2">&quot;your_graph_data.pt&quot;</span>
<span class="n">feature_data</span> <span class="o">=</span> <span class="s2">&quot;your_feature_matrix.h5ad&quot;</span>
<span class="c1">## save the output adata, default is set to the current folder</span>
<span class="n">output_path</span> <span class="o">=</span> <span class="s1">&#39;./&#39;</span>
</pre></div>
</div>
</div>
</section>
<section id="Set-Hyperparameters-and-Train-the-Model">
<h2>Set Hyperparameters and Train the Model<a class="headerlink" href="#Set-Hyperparameters-and-Train-the-Model" title="Permalink to this heading"></a></h2>
<p>The function <code class="docutils literal notranslate"><span class="pre">run_DOLPHIN</span></code> is used to configure hyperparameters and train the model. Below is a detailed explanation of its parameters:</p>
<hr class="docutils" />
<section id="Function-Definition">
<h3><strong>Function Definition</strong><a class="headerlink" href="#Function-Definition" title="Permalink to this heading"></a></h3>
<p>```python run_DOLPHIN(data_type, graph_in, fea_in, current_out_path=’./’, params=None, device=’cuda:0’, seed_num=0)</p>
<section id="Parameters">
<h4>Parameters<a class="headerlink" href="#Parameters" title="Permalink to this heading"></a></h4>
<section id="1.-data_type-Specifies-the-type-of-input-single-cell-RNA-seq-data:">
<h5>1. <code class="docutils literal notranslate"><span class="pre">data_type</span></code> Specifies the type of input single-cell RNA-seq data:<a class="headerlink" href="#1.-data_type-Specifies-the-type-of-input-single-cell-RNA-seq-data:" title="Permalink to this heading"></a></h5>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;full-length&quot;</span></code>: For full-length RNA-seq data.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;10x&quot;</span></code>: For 10x Genomics RNA-seq data.</p></li>
</ul>
</section>
<section id="2.-graph_in-The-input-graph-dataset.">
<h5>2. <code class="docutils literal notranslate"><span class="pre">graph_in</span></code> The input graph dataset.<a class="headerlink" href="#2.-graph_in-The-input-graph-dataset." title="Permalink to this heading"></a></h5>
</section>
<section id="3.-fea_in-The-input-feature-matrix,-provided-as-an-AnnData-object-(adata).">
<h5>3. <code class="docutils literal notranslate"><span class="pre">fea_in</span></code> The input feature matrix, provided as an AnnData object (<code class="docutils literal notranslate"><span class="pre">adata</span></code>).<a class="headerlink" href="#3.-fea_in-The-input-feature-matrix,-provided-as-an-AnnData-object-(adata)." title="Permalink to this heading"></a></h5>
</section>
<section id="4.-current_out_path-Specifies-the-output-directory-where-the-resulting-cell-embeddings-(X_z)-will-be-saved.-The-output-file-will-be-named:-DOLPHIN_Z.h5ad">
<h5>4. <code class="docutils literal notranslate"><span class="pre">current_out_path</span></code> Specifies the output directory where the resulting cell embeddings (<code class="docutils literal notranslate"><span class="pre">X_z</span></code>) will be saved. The output file will be named: <code class="docutils literal notranslate"><span class="pre">DOLPHIN_Z.h5ad</span></code><a class="headerlink" href="#4.-current_out_path-Specifies-the-output-directory-where-the-resulting-cell-embeddings-(X_z)-will-be-saved.-The-output-file-will-be-named:-DOLPHIN_Z.h5ad" title="Permalink to this heading"></a></h5>
</section>
<section id="5.-params-Model-hyperparameters.">
<h5>5. <code class="docutils literal notranslate"><span class="pre">params</span></code> Model hyperparameters.<a class="headerlink" href="#5.-params-Model-hyperparameters." title="Permalink to this heading"></a></h5>
<div class="line-block">
<div class="line">If <code class="docutils literal notranslate"><span class="pre">data_type</span></code> is set, you can use the <strong>default hyperparameters</strong> or provide your own in a dictionary format.</div>
<div class="line">Below is a list of customizable hyperparameters:</div>
</div>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Parameter</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">&quot;gat_channel&quot;</span></code></p></td>
<td><p>Number of features per node after the GAT layer.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">&quot;nhead&quot;</span></code></p></td>
<td><p>Number of attention heads in the graph attention layer.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">&quot;gat_dropout&quot;</span></code></p></td>
<td><p>Dropout rate for the GAT layer.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">&quot;list_gra_enc_hid&quot;</span></code></p></td>
<td><p>Neuron sizes for each fully connected layer of the encoder.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">&quot;gra_p_dropout&quot;</span></code></p></td>
<td><p>Dropout rate for the encoder.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">&quot;z_dim&quot;</span></code></p></td>
<td><p>Dimensionality of the latent Z space.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">&quot;list_fea_dec_hid&quot;</span></code></p></td>
<td><p>Neuron sizes for each fully connected layer of the feature decoder.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">&quot;list_adj_dec_hid&quot;</span></code></p></td>
<td><p>Neuron sizes for each fully connected layer of the adjacency decoder.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">&quot;lr&quot;</span></code></p></td>
<td><p>Learning rate for optimization.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">&quot;batch&quot;</span></code></p></td>
<td><p>Mini-batch size.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">&quot;epochs&quot;</span></code></p></td>
<td><p>Number of training epochs.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">&quot;kl_beta&quot;</span></code></p></td>
<td><p>KL divergence weight.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">&quot;fea_lambda&quot;</span></code></p></td>
<td><p>Feature matrix reconstruction loss weight.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">&quot;adj_lambda&quot;</span></code></p></td>
<td><p>Adjacency matrix reconstruction loss weight.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="6.-device-Specifies-the-device-for-training.-Default:-%22cuda:0%22-for-GPU-training-(highly-recommended).">
<h5>6. <code class="docutils literal notranslate"><span class="pre">device</span></code> Specifies the device for training. Default: <code class="docutils literal notranslate"><span class="pre">&quot;cuda:0&quot;</span></code> for GPU training (highly recommended).<a class="headerlink" href="#6.-device-Specifies-the-device-for-training.-Default:-%22cuda:0%22-for-GPU-training-(highly-recommended)." title="Permalink to this heading"></a></h5>
</section>
<section id="7.-seed_numSets-the-random-seed-for-reproducibility.">
<h5>7. <code class="docutils literal notranslate"><span class="pre">seed_num</span></code>Sets the random seed for reproducibility.<a class="headerlink" href="#7.-seed_numSets-the-random-seed-for-reproducibility." title="Permalink to this heading"></a></h5>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">run_DOLPHIN</span><span class="p">(</span><span class="s2">&quot;full-length&quot;</span><span class="p">,</span> <span class="n">graph_data</span><span class="p">,</span> <span class="n">feature_data</span><span class="p">,</span> <span class="n">output_path</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
</section>
</section>
</section>
<section id="Cell-Embedding-Cluster-Using-X_z">
<h2>Cell Embedding Cluster Using <code class="docutils literal notranslate"><span class="pre">X_z</span></code><a class="headerlink" href="#Cell-Embedding-Cluster-Using-X_z" title="Permalink to this heading"></a></h2>
<div class="line-block">
<div class="line">The cell embedding matrix <code class="docutils literal notranslate"><span class="pre">X_z</span></code> represents the low-dimensional latent space learned by the DOLPHIN model.</div>
<div class="line">This matrix can be used to visualize cell clusters and analyze their relationships in the latent space.</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">from</span> <span class="nn">sklearn.metrics</span> <span class="kn">import</span> <span class="n">adjusted_rand_score</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s2">&quot;./DOLPHIN_Z.h5ad&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">use_rep</span><span class="o">=</span><span class="s2">&quot;X_z&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">leiden</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;leiden&quot;</span><span class="p">])))</span>
<span class="n">adjusted_rand_score</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;celltype&quot;</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;leiden&quot;</span><span class="p">])</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;leiden&#39;</span><span class="p">,</span> <span class="s2">&quot;celltype&quot;</span><span class="p">],</span> <span class="n">wspace</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="step2_5_graph_data_generation.html" class="btn btn-neutral float-left" title="Generate Graph Data for MODEl input" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="step4_cell_aggregation.html" class="btn btn-neutral float-right" title="Cell Aggregation" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright Ding Lab at McGill University.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>