<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Exon-Level Reference GTF File Generation &mdash; DOLPHIN  documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/copybutton.css" type="text/css" />
      <link rel="stylesheet" href="../_static/nbsphinx-code-cells.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/jquery.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
        <script src="../_static/clipboard.min.js"></script>
        <script src="../_static/copybutton.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Preprocessing Full-Length Single-Cell RNA-Seq for Exon and Junction Read Counts" href="step1_1_preprocess_full_length.html" />
    <link rel="prev" title="Tutorials" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            DOLPHIN
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Getting started with DOLPHIN:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../usage.html">Usage</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Tutorials</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">Exon-Level Reference GTF File Generation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#For-Human-GRCh38"><strong>For Human GRCh38</strong></a></li>
<li class="toctree-l3"><a class="reference internal" href="#For-Other-Species"><strong>For Other Species</strong></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="step1_1_preprocess_full_length.html">Preprocessing Full-Length Single-Cell RNA-Seq for Exon and Junction Read Counts</a></li>
<li class="toctree-l2"><a class="reference internal" href="step1_2_preprocess_10X.html">Preprocessing 10X Single-Cell RNA-Seq for Exon and Junction Read Counts</a></li>
<li class="toctree-l2"><a class="reference internal" href="step2_1_graph_generation.html">Generate Feature Matrix and Adjacency Matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="step2_2_exonGene_matrix_generation.html">Generate Exon Gene h5ad and Select Highly Variable Genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="step2_3_feature_matrix_generation.html">Generate Feature Matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="step2_4_adjacency_matrix_generation.html">Generate Adjacency Matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="step2_5_graph_data_generation.html">Generate Graph Data for MODEl input</a></li>
<li class="toctree-l2"><a class="reference internal" href="step3_run_DOLPHIN.html">Setup and Train the DOLPHIN Model</a></li>
<li class="toctree-l2"><a class="reference internal" href="step4_cell_aggregation.html">Cell Aggregation</a></li>
<li class="toctree-l2"><a class="reference internal" href="step5_alternative_splicing.html">Detect Alternative Splicing Events with Outrigger</a></li>
<li class="toctree-l2"><a class="reference internal" href="step6_alternative_splicing_analysis.html">Alternative Splicing Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="step6_alternative_splicing_analysis.html#Generate-PSI-adata">Generate PSI adata</a></li>
<li class="toctree-l2"><a class="reference internal" href="step6_alternative_splicing_analysis.html#Differential-Alternative-Splicing-Events">Differential Alternative Splicing Events</a></li>
<li class="toctree-l2"><a class="reference internal" href="step7_1_MAST.html">Run MAST to get Exon level p-value</a></li>
<li class="toctree-l2"><a class="reference internal" href="step7_2_EDEG_analysis.html">Exon-Level Differential Gene Analysis Using Feature Matrix</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Examples</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../examples/index.html">Examples</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API Documentations</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../API/index.html">API Documentations</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">DOLPHIN</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">Tutorials</a></li>
      <li class="breadcrumb-item active">Exon-Level Reference GTF File Generation</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/tutorials/step0_generate_exon_gtf.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="Exon-Level-Reference-GTF-File-Generation">
<h1>Exon-Level Reference GTF File Generation<a class="headerlink" href="#Exon-Level-Reference-GTF-File-Generation" title="Permalink to this heading"></a></h1>
<p>This guide explains how to generate an exon-level GTF reference file. This file is used to align scRNA-seq data to the exon level, allowing the extraction of exon read counts and junction read counts. The goal of the exon-level GTF is to ensure that exons within each gene are unique and do not overlap with one another.</p>
<p><img alt="This is an example image" src="../_images/exon_gtf_demonstration.png" /></p>
<section id="For-Human-GRCh38">
<h2><strong>For Human GRCh38</strong><a class="headerlink" href="#For-Human-GRCh38" title="Permalink to this heading"></a></h2>
<p>You can directly download the pre-generated exon-level GTF file from <a class="reference external" href="https://mcgill-my.sharepoint.com/my?id=%2Fpersonal%2Fkailu%5Fsong%5Fmail%5Fmcgill%5Fca%2FDocuments%2FDeepExonas%5Fgithub%5Fexample%2Falignment%5Ffiles">here</a>.</p>
</section>
<section id="For-Other-Species">
<h2><strong>For Other Species</strong><a class="headerlink" href="#For-Other-Species" title="Permalink to this heading"></a></h2>
<ol class="arabic simple">
<li><p>First, download the reference GTF file from <a class="reference external" href="https://www.ensembl.org/index.html">here</a>.</p></li>
<li><p>Then, run this script to generate the exon-level GTF file.</p></li>
</ol>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#import annotation file</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">DOLPHIN</span> <span class="kn">import</span> <span class="n">gtfpy</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">math</span>

<span class="n">pd</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="s1">&#39;display.max_columns&#39;</span><span class="p">,</span><span class="mi">500</span><span class="p">)</span>
<span class="n">pd</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="s1">&#39;display.max_rows&#39;</span><span class="p">,</span><span class="mi">100</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/mnt/md0/kailu/anaconda3/envs/DOLPHIN/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#load the original ensembl gtf file</span>
<span class="n">df_gtf</span><span class="o">=</span><span class="n">gtfpy</span><span class="o">.</span><span class="n">readGTF</span><span class="p">(</span><span class="s2">&quot;Homo_sapiens.GRCh38.107.gtf&quot;</span><span class="p">)</span>
<span class="n">GTFpa</span><span class="o">=</span><span class="n">gtfpy</span><span class="o">.</span><span class="n">parseGTF</span><span class="p">(</span><span class="n">df_gtf</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">GTFpa</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>seqname</th>
      <th>source</th>
      <th>feature</th>
      <th>start</th>
      <th>end</th>
      <th>score</th>
      <th>strand</th>
      <th>frame</th>
      <th>gene_id</th>
      <th>protein_id</th>
      <th>gene_source</th>
      <th>exon_number</th>
      <th>gene_biotype</th>
      <th>transcript_version</th>
      <th>transcript_support_level</th>
      <th>exon_version</th>
      <th>exon_id</th>
      <th>transcript_source</th>
      <th>transcript_name</th>
      <th>gene_name</th>
      <th>tag</th>
      <th>ccds_id</th>
      <th>transcript_biotype</th>
      <th>transcript_id</th>
      <th>protein_version</th>
      <th>gene_version</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>1</td>
      <td>ensembl_havana</td>
      <td>gene</td>
      <td>1471765</td>
      <td>1497848</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ENSG00000160072</td>
      <td>NaN</td>
      <td>ensembl_havana</td>
      <td>NaN</td>
      <td>protein_coding</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>ATAD3B</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>20</td>
    </tr>
    <tr>
      <th>1</th>
      <td>1</td>
      <td>ensembl_havana</td>
      <td>transcript</td>
      <td>1471765</td>
      <td>1497848</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ENSG00000160072</td>
      <td>NaN</td>
      <td>ensembl_havana</td>
      <td>NaN</td>
      <td>protein_coding</td>
      <td>1</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>ensembl_havana</td>
      <td>ATAD3B-206</td>
      <td>ATAD3B</td>
      <td>basic</td>
      <td>CCDS30</td>
      <td>protein_coding</td>
      <td>ENST00000673477</td>
      <td>NaN</td>
      <td>20</td>
    </tr>
    <tr>
      <th>2</th>
      <td>1</td>
      <td>ensembl_havana</td>
      <td>exon</td>
      <td>1471765</td>
      <td>1472089</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ENSG00000160072</td>
      <td>NaN</td>
      <td>ensembl_havana</td>
      <td>1</td>
      <td>protein_coding</td>
      <td>1</td>
      <td>NaN</td>
      <td>1</td>
      <td>ENSE00003889014</td>
      <td>ensembl_havana</td>
      <td>ATAD3B-206</td>
      <td>ATAD3B</td>
      <td>basic</td>
      <td>CCDS30</td>
      <td>protein_coding</td>
      <td>ENST00000673477</td>
      <td>NaN</td>
      <td>20</td>
    </tr>
    <tr>
      <th>3</th>
      <td>1</td>
      <td>ensembl_havana</td>
      <td>CDS</td>
      <td>1471885</td>
      <td>1472089</td>
      <td>.</td>
      <td>+</td>
      <td>0</td>
      <td>ENSG00000160072</td>
      <td>ENSP00000500094</td>
      <td>ensembl_havana</td>
      <td>1</td>
      <td>protein_coding</td>
      <td>1</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>ensembl_havana</td>
      <td>ATAD3B-206</td>
      <td>ATAD3B</td>
      <td>basic</td>
      <td>CCDS30</td>
      <td>protein_coding</td>
      <td>ENST00000673477</td>
      <td>1</td>
      <td>20</td>
    </tr>
    <tr>
      <th>4</th>
      <td>1</td>
      <td>ensembl_havana</td>
      <td>start_codon</td>
      <td>1471885</td>
      <td>1471887</td>
      <td>.</td>
      <td>+</td>
      <td>0</td>
      <td>ENSG00000160072</td>
      <td>NaN</td>
      <td>ensembl_havana</td>
      <td>1</td>
      <td>protein_coding</td>
      <td>1</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>ensembl_havana</td>
      <td>ATAD3B-206</td>
      <td>ATAD3B</td>
      <td>basic</td>
      <td>CCDS30</td>
      <td>protein_coding</td>
      <td>ENST00000673477</td>
      <td>NaN</td>
      <td>20</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>3371239</th>
      <td>KI270713.1</td>
      <td>ensembl</td>
      <td>five_prime_utr</td>
      <td>32373</td>
      <td>32528</td>
      <td>.</td>
      <td>-</td>
      <td>.</td>
      <td>ENSG00000277475</td>
      <td>NaN</td>
      <td>ensembl</td>
      <td>NaN</td>
      <td>protein_coding</td>
      <td>1</td>
      <td>NA</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>ensembl</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>basic</td>
      <td>NaN</td>
      <td>protein_coding</td>
      <td>ENST00000612315</td>
      <td>NaN</td>
      <td>1</td>
    </tr>
    <tr>
      <th>3371240</th>
      <td>KI270713.1</td>
      <td>ensembl</td>
      <td>three_prime_utr</td>
      <td>31698</td>
      <td>31841</td>
      <td>.</td>
      <td>-</td>
      <td>.</td>
      <td>ENSG00000277475</td>
      <td>NaN</td>
      <td>ensembl</td>
      <td>NaN</td>
      <td>protein_coding</td>
      <td>1</td>
      <td>NA</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>ensembl</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>basic</td>
      <td>NaN</td>
      <td>protein_coding</td>
      <td>ENST00000612315</td>
      <td>NaN</td>
      <td>1</td>
    </tr>
    <tr>
      <th>3371241</th>
      <td>KI270713.1</td>
      <td>ensembl</td>
      <td>gene</td>
      <td>21861</td>
      <td>22024</td>
      <td>.</td>
      <td>-</td>
      <td>.</td>
      <td>ENSG00000275405</td>
      <td>NaN</td>
      <td>ensembl</td>
      <td>NaN</td>
      <td>snRNA</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>U1</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>1</td>
    </tr>
    <tr>
      <th>3371242</th>
      <td>KI270713.1</td>
      <td>ensembl</td>
      <td>transcript</td>
      <td>21861</td>
      <td>22024</td>
      <td>.</td>
      <td>-</td>
      <td>.</td>
      <td>ENSG00000275405</td>
      <td>NaN</td>
      <td>ensembl</td>
      <td>NaN</td>
      <td>snRNA</td>
      <td>1</td>
      <td>NA</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>ensembl</td>
      <td>U1.26-201</td>
      <td>U1</td>
      <td>basic</td>
      <td>NaN</td>
      <td>snRNA</td>
      <td>ENST00000619109</td>
      <td>NaN</td>
      <td>1</td>
    </tr>
    <tr>
      <th>3371243</th>
      <td>KI270713.1</td>
      <td>ensembl</td>
      <td>exon</td>
      <td>21861</td>
      <td>22024</td>
      <td>.</td>
      <td>-</td>
      <td>.</td>
      <td>ENSG00000275405</td>
      <td>NaN</td>
      <td>ensembl</td>
      <td>1</td>
      <td>snRNA</td>
      <td>1</td>
      <td>NA</td>
      <td>1</td>
      <td>ENSE00003747303</td>
      <td>ensembl</td>
      <td>U1.26-201</td>
      <td>U1</td>
      <td>basic</td>
      <td>NaN</td>
      <td>snRNA</td>
      <td>ENST00000619109</td>
      <td>NaN</td>
      <td>1</td>
    </tr>
  </tbody>
</table>
<p>3371244 rows × 26 columns</p>
</div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># modify gtf file which only keep unique exon</span>
<span class="n">df_exon</span> <span class="o">=</span> <span class="n">GTFpa</span><span class="p">[(</span><span class="n">GTFpa</span><span class="p">[</span><span class="s2">&quot;feature&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;exon&quot;</span><span class="p">)][[</span><span class="s1">&#39;seqname&#39;</span><span class="p">,</span> <span class="s1">&#39;source&#39;</span><span class="p">,</span><span class="s1">&#39;feature&#39;</span><span class="p">,</span><span class="s1">&#39;start&#39;</span><span class="p">,</span><span class="s1">&#39;end&#39;</span><span class="p">,</span><span class="s1">&#39;score&#39;</span><span class="p">,</span><span class="s1">&#39;strand&#39;</span><span class="p">,</span><span class="s1">&#39;frame&#39;</span><span class="p">,</span><span class="s1">&#39;gene_id&#39;</span><span class="p">,</span><span class="s1">&#39;gene_version&#39;</span><span class="p">,</span><span class="s1">&#39;gene_name&#39;</span><span class="p">,</span><span class="s1">&#39;gene_source&#39;</span><span class="p">,</span><span class="s1">&#39;gene_biotype&#39;</span><span class="p">,</span><span class="s1">&#39;exon_number&#39;</span><span class="p">]]</span>
<span class="n">df_exon</span><span class="p">[[</span><span class="s2">&quot;start&quot;</span><span class="p">,</span><span class="s2">&quot;end&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="n">df_exon</span><span class="p">[[</span><span class="s2">&quot;start&quot;</span><span class="p">,</span><span class="s2">&quot;end&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">to_numeric</span><span class="p">)</span>
<span class="n">df_exon</span><span class="o">=</span><span class="n">df_exon</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;seqname&quot;</span><span class="p">,</span> <span class="s2">&quot;start&quot;</span><span class="p">,</span><span class="s2">&quot;end&quot;</span><span class="p">],</span><span class="n">ascending</span><span class="o">=</span><span class="p">[</span><span class="kc">True</span><span class="p">,</span><span class="kc">True</span><span class="p">,</span><span class="kc">True</span><span class="p">])</span>
<span class="c1">#1. remove duplicate exon within gene which has same start and end site.</span>
<span class="n">df_exon_nodup</span><span class="o">=</span><span class="n">df_exon</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;gene_id&quot;</span><span class="p">,</span><span class="s2">&quot;start&quot;</span><span class="p">,</span><span class="s2">&quot;end&quot;</span><span class="p">],</span><span class="n">keep</span><span class="o">=</span><span class="s2">&quot;first&quot;</span><span class="p">)</span>
<span class="n">df_exon_nodup</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">50</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#2 get unique location for exon</span>
<span class="kn">import</span> <span class="nn">math</span>

<span class="k">def</span> <span class="nf">exon_uniq</span><span class="p">(</span><span class="n">gene</span><span class="p">):</span>
    <span class="n">_df_exon</span><span class="o">=</span><span class="n">df_exon_nodup</span><span class="o">.</span><span class="n">loc</span><span class="p">[(</span><span class="n">df_exon_nodup</span><span class="p">[</span><span class="s2">&quot;gene_id&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">gene</span><span class="p">)]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

    <span class="c1"># save the start/end sites in list</span>
    <span class="n">_exon</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">_df_exon</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
        <span class="n">_exon</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">_df_exon</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="s2">&quot;start&quot;</span><span class="p">],</span><span class="n">_df_exon</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="s2">&quot;end&quot;</span><span class="p">]])</span>

    <span class="c1">#create a unique list to store unique location</span>
    <span class="n">_unq</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">_unq</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">_exon</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

    <span class="c1">#get unique list of exon site:</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">_exon</span><span class="p">)):</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">_unq</span><span class="p">)):</span>
            <span class="k">if</span> <span class="n">_exon</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="nb">max</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">max</span><span class="p">,</span> <span class="n">_unq</span><span class="p">)):</span>
                <span class="n">_unq</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">_exon</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="p">(</span><span class="n">_unq</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">_exon</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">_unq</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="mi">1</span><span class="p">])</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">_exon</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">_unq</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="mi">1</span><span class="p">]):</span>
                    <span class="n">_unq</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">_exon</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>

    <span class="k">def</span> <span class="nf">unique</span><span class="p">(</span><span class="n">list1</span><span class="p">):</span>
        <span class="n">unique_list</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">list1</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">x</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">unique_list</span><span class="p">:</span>
                <span class="n">unique_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
        <span class="k">return</span><span class="p">(</span><span class="n">unique_list</span><span class="p">)</span>

    <span class="n">_unq</span> <span class="o">=</span> <span class="n">unique</span><span class="p">(</span><span class="n">_unq</span><span class="p">)</span>

    <span class="c1">#process the exon table to only keep unique exon location, double confirm if all the cases are considered.</span>
    <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">_df_exon</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">_unq</span><span class="p">)):</span>
            <span class="k">if</span> <span class="p">((</span><span class="n">_unq</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">_df_exon</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">k</span><span class="p">][</span><span class="s1">&#39;start&#39;</span><span class="p">])</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">_unq</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">_df_exon</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">k</span><span class="p">][</span><span class="s1">&#39;end&#39;</span><span class="p">])):</span>
                <span class="n">_df_exon</span><span class="o">.</span><span class="n">at</span><span class="p">[</span><span class="n">k</span><span class="p">,</span><span class="s1">&#39;_start&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">_unq</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">_df_exon</span><span class="o">.</span><span class="n">at</span><span class="p">[</span><span class="n">k</span><span class="p">,</span><span class="s1">&#39;_end&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">_unq</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">_df_exon</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
        <span class="k">if</span> <span class="n">math</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">_df_exon</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="s1">&#39;_start&#39;</span><span class="p">])</span> <span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Attention: &quot;</span> <span class="o">+</span> <span class="n">_df_exon</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="s1">&#39;gene_id&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;,Start=&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">_df_exon</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="s1">&#39;start&#39;</span><span class="p">])</span> <span class="o">+</span> <span class="s2">&quot;,end=&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">_df_exon</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="s1">&#39;end&#39;</span><span class="p">])</span> <span class="o">+</span> <span class="s2">&quot; is not assigned value&quot;</span><span class="p">)</span>

    <span class="c1">#only keep unique exon and clean the datasets</span>
    <span class="n">_df_exon_out</span><span class="o">=</span><span class="n">_df_exon</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;_start&quot;</span><span class="p">,</span><span class="s2">&quot;_end&quot;</span><span class="p">],</span><span class="n">keep</span><span class="o">=</span><span class="s2">&quot;first&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">_df_exon_out</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">_df_exon_out</span><span class="p">[</span><span class="s1">&#39;_start&#39;</span><span class="p">]</span>
    <span class="n">_df_exon_out</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">_df_exon_out</span><span class="p">[</span><span class="s1">&#39;_end&#39;</span><span class="p">]</span>

    <span class="n">_df_exon_out</span><span class="p">[[</span><span class="s2">&quot;start&quot;</span><span class="p">,</span><span class="s2">&quot;end&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="n">_df_exon_out</span><span class="p">[[</span><span class="s2">&quot;start&quot;</span><span class="p">,</span><span class="s2">&quot;end&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;int&#39;</span><span class="p">)</span>
    <span class="n">_df_exon_out</span> <span class="o">=</span> <span class="n">_df_exon_out</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;_start&quot;</span><span class="p">,</span><span class="s2">&quot;_end&quot;</span><span class="p">])</span>

    <span class="c1">#recount exon number value</span>
    <span class="n">_df_exon_out</span> <span class="o">=</span> <span class="n">_df_exon_out</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
    <span class="n">_df_exon_out</span><span class="p">[</span><span class="s1">&#39;exon_number&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">_df_exon_out</span><span class="o">.</span><span class="n">index</span><span class="o">+</span><span class="mi">1</span>

    <span class="k">return</span><span class="p">(</span><span class="n">_df_exon_out</span><span class="p">)</span>

<span class="c1">#check function</span>
<span class="n">exon_uniq</span><span class="p">(</span><span class="s2">&quot;ENSG00000160072&quot;</span><span class="p">)</span>
<span class="n">exon_uniq</span><span class="p">(</span><span class="s2">&quot;ENSG00000234396&quot;</span><span class="p">)</span>
<span class="n">exon_uniq</span><span class="p">(</span><span class="s2">&quot;ENSG00000223972&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Run genes in batches and save each DataFrame into a separate <code class="docutils literal notranslate"><span class="pre">.pkl</span></code> file. A new <code class="docutils literal notranslate"><span class="pre">.pkl</span></code> file will be created for every 10,000 genes.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># get the unique exon location per gene</span>
<span class="n">gene_list</span> <span class="o">=</span> <span class="n">df_exon_nodup</span><span class="o">.</span><span class="n">gene_id</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="n">saveNum</span> <span class="o">=</span> <span class="mi">10000</span>
<span class="n">list_of_df</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>
<span class="c1">#run gene in batch, save each df into a pkl file</span>
<span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="nb">int</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gene_list</span><span class="p">)</span><span class="o">/</span><span class="n">saveNum</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span><span class="p">):</span>
    <span class="c1">#initialize output dataframe</span>
    <span class="n">df_out</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
    <span class="c1">#get the unique gene_id list</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">j</span><span class="o">*</span><span class="n">saveNum</span><span class="p">,</span><span class="n">j</span><span class="o">*</span><span class="n">saveNum</span><span class="o">+</span><span class="n">saveNum</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">i</span> <span class="o">&gt;=</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene_list</span><span class="p">):</span>
            <span class="k">break</span>
        <span class="n">_temp</span> <span class="o">=</span> <span class="n">exon_uniq</span><span class="p">(</span><span class="n">gene_list</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
        <span class="n">df_out</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">df_out</span><span class="p">,</span> <span class="n">_temp</span><span class="p">],</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">gene_list</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
    <span class="n">list_of_df</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">df_out</span><span class="p">)</span>
    <span class="n">df_out</span><span class="o">.</span><span class="n">to_pickle</span><span class="p">(</span><span class="s2">&quot;df_exon_gtf_&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">j</span><span class="p">)</span><span class="o">+</span><span class="s2">&quot;.pkl&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>

<span class="n">pkl_files</span> <span class="o">=</span> <span class="p">[</span><span class="n">file</span> <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)</span> <span class="k">if</span> <span class="n">file</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;df_exon_gtf_&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">file</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;.pkl&#39;</span><span class="p">)]</span>

<span class="n">dataframes</span> <span class="o">=</span> <span class="p">[</span><span class="n">pd</span><span class="o">.</span><span class="n">read_pickle</span><span class="p">(</span><span class="n">file</span><span class="p">)</span> <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">pkl_files</span><span class="p">]</span>
<span class="n">gtf_all</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">dataframes</span><span class="p">,</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># Check the result</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Successfully combined </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">pkl_files</span><span class="p">)</span><span class="si">}</span><span class="s2"> files into a single DataFrame.&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">gtf_all</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>
<br/></pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#check df_out has identitcal locations</span>
<span class="n">_ck_out</span> <span class="o">=</span> <span class="n">gtf_all</span>
<span class="c1"># _ck_out = df_out.copy()</span>
<span class="n">_ck_out</span><span class="p">[</span><span class="s1">&#39;_next_start&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">_ck_out</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;gene_id&#39;</span><span class="p">)[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">shift</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>
<span class="n">_ck_out</span><span class="p">[</span><span class="s1">&#39;_ck&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">_ck_out</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
    <span class="k">if</span> <span class="n">math</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">_ck_out</span><span class="p">[</span><span class="s1">&#39;_next_start&#39;</span><span class="p">][</span><span class="n">i</span><span class="p">]):</span>
        <span class="n">_ck_out</span><span class="p">[</span><span class="s2">&quot;_ck&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">_ck_out</span><span class="p">[</span><span class="s2">&quot;_ck&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">_ck_out</span><span class="p">[</span><span class="s1">&#39;_next_start&#39;</span><span class="p">][</span><span class="n">i</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">_ck_out</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">][</span><span class="n">i</span><span class="p">]</span>

<span class="n">_ck_out</span> <span class="o">=</span> <span class="n">_ck_out</span><span class="p">[(</span><span class="n">_ck_out</span><span class="p">[</span><span class="s2">&quot;_ck&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="kc">False</span><span class="p">)]</span>
<span class="n">_ck_out</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[111]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><br/><span></span><span class="n">gtf_all</span> <span class="o">=</span> <span class="n">gtf_all</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;level_0&#39;</span><span class="p">,</span><span class="s1">&#39;index&#39;</span><span class="p">])</span>
<span class="n">gtf_all</span><span class="o">=</span><span class="n">gtf_all</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;seqname&quot;</span><span class="p">,</span> <span class="s2">&quot;gene_id&quot;</span><span class="p">,</span> <span class="s2">&quot;start&quot;</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">],</span><span class="n">ascending</span><span class="o">=</span><span class="p">[</span><span class="kc">True</span><span class="p">,</span><span class="kc">True</span><span class="p">,</span><span class="kc">True</span><span class="p">,</span><span class="kc">True</span><span class="p">])</span>
<span class="n">gtf_all</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[111]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>seqname</th>
      <th>source</th>
      <th>feature</th>
      <th>start</th>
      <th>end</th>
      <th>score</th>
      <th>strand</th>
      <th>frame</th>
      <th>gene_id</th>
      <th>gene_version</th>
      <th>gene_name</th>
      <th>gene_source</th>
      <th>gene_biotype</th>
      <th>exon_number</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>6939</th>
      <td>1</td>
      <td>ensembl_havana</td>
      <td>exon</td>
      <td>169849631</td>
      <td>169853772</td>
      <td>.</td>
      <td>-</td>
      <td>.</td>
      <td>ENSG00000000457</td>
      <td>14</td>
      <td>SCYL3</td>
      <td>ensembl_havana</td>
      <td>protein_coding</td>
      <td>1</td>
    </tr>
    <tr>
      <th>6940</th>
      <td>1</td>
      <td>ensembl_havana</td>
      <td>exon</td>
      <td>169854270</td>
      <td>169854964</td>
      <td>.</td>
      <td>-</td>
      <td>.</td>
      <td>ENSG00000000457</td>
      <td>14</td>
      <td>SCYL3</td>
      <td>ensembl_havana</td>
      <td>protein_coding</td>
      <td>2</td>
    </tr>
    <tr>
      <th>6941</th>
      <td>1</td>
      <td>havana</td>
      <td>exon</td>
      <td>169855796</td>
      <td>169855957</td>
      <td>.</td>
      <td>-</td>
      <td>.</td>
      <td>ENSG00000000457</td>
      <td>14</td>
      <td>SCYL3</td>
      <td>ensembl_havana</td>
      <td>protein_coding</td>
      <td>3</td>
    </tr>
    <tr>
      <th>6942</th>
      <td>1</td>
      <td>ensembl_havana</td>
      <td>exon</td>
      <td>169859041</td>
      <td>169859212</td>
      <td>.</td>
      <td>-</td>
      <td>.</td>
      <td>ENSG00000000457</td>
      <td>14</td>
      <td>SCYL3</td>
      <td>ensembl_havana</td>
      <td>protein_coding</td>
      <td>4</td>
    </tr>
    <tr>
      <th>6943</th>
      <td>1</td>
      <td>ensembl_havana</td>
      <td>exon</td>
      <td>169862613</td>
      <td>169862797</td>
      <td>.</td>
      <td>-</td>
      <td>.</td>
      <td>ENSG00000000457</td>
      <td>14</td>
      <td>SCYL3</td>
      <td>ensembl_havana</td>
      <td>protein_coding</td>
      <td>5</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>341415</th>
      <td>Y</td>
      <td>havana</td>
      <td>exon</td>
      <td>21250550</td>
      <td>21250708</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ENSG00000289707</td>
      <td>1</td>
      <td>NaN</td>
      <td>havana</td>
      <td>lncRNA</td>
      <td>8</td>
    </tr>
    <tr>
      <th>341416</th>
      <td>Y</td>
      <td>havana</td>
      <td>exon</td>
      <td>21253123</td>
      <td>21253193</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ENSG00000289707</td>
      <td>1</td>
      <td>NaN</td>
      <td>havana</td>
      <td>lncRNA</td>
      <td>9</td>
    </tr>
    <tr>
      <th>341417</th>
      <td>Y</td>
      <td>havana</td>
      <td>exon</td>
      <td>21255420</td>
      <td>21257832</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ENSG00000289707</td>
      <td>1</td>
      <td>NaN</td>
      <td>havana</td>
      <td>lncRNA</td>
      <td>10</td>
    </tr>
    <tr>
      <th>341902</th>
      <td>Y</td>
      <td>havana_tagene</td>
      <td>exon</td>
      <td>7274826</td>
      <td>7275065</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ENSG00000289826</td>
      <td>1</td>
      <td>NaN</td>
      <td>havana_tagene</td>
      <td>lncRNA</td>
      <td>1</td>
    </tr>
    <tr>
      <th>341903</th>
      <td>Y</td>
      <td>havana_tagene</td>
      <td>exon</td>
      <td>7290515</td>
      <td>7290751</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ENSG00000289826</td>
      <td>1</td>
      <td>NaN</td>
      <td>havana_tagene</td>
      <td>lncRNA</td>
      <td>2</td>
    </tr>
  </tbody>
</table>
<p>354393 rows × 14 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[109]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#manual check</span>
<span class="n">gtf_all</span><span class="p">[(</span><span class="n">gtf_all</span><span class="p">[</span><span class="s2">&quot;gene_id&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;ENSG00000223972&quot;</span><span class="p">)]</span>
<span class="n">GTFpa</span><span class="p">[(</span><span class="n">GTFpa</span><span class="p">[</span><span class="s2">&quot;gene_id&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;ENSG00000223972&quot;</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">GTFpa</span><span class="p">[</span><span class="s2">&quot;feature&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;exon&quot;</span><span class="p">)]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[109]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>seqname</th>
      <th>source</th>
      <th>feature</th>
      <th>start</th>
      <th>end</th>
      <th>score</th>
      <th>strand</th>
      <th>frame</th>
      <th>gene_id</th>
      <th>gene_version</th>
      <th>gene_name</th>
      <th>gene_source</th>
      <th>gene_biotype</th>
      <th>exon_number</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>20024</th>
      <td>1</td>
      <td>havana</td>
      <td>exon</td>
      <td>11869</td>
      <td>12227</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ENSG00000223972</td>
      <td>5</td>
      <td>DDX11L1</td>
      <td>havana</td>
      <td>transcribed_unprocessed_pseudogene</td>
      <td>1</td>
    </tr>
    <tr>
      <th>20025</th>
      <td>1</td>
      <td>havana</td>
      <td>exon</td>
      <td>12613</td>
      <td>12721</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ENSG00000223972</td>
      <td>5</td>
      <td>DDX11L1</td>
      <td>havana</td>
      <td>transcribed_unprocessed_pseudogene</td>
      <td>2</td>
    </tr>
    <tr>
      <th>20026</th>
      <td>1</td>
      <td>havana</td>
      <td>exon</td>
      <td>12975</td>
      <td>13052</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ENSG00000223972</td>
      <td>5</td>
      <td>DDX11L1</td>
      <td>havana</td>
      <td>transcribed_unprocessed_pseudogene</td>
      <td>3</td>
    </tr>
    <tr>
      <th>20027</th>
      <td>1</td>
      <td>havana</td>
      <td>exon</td>
      <td>13221</td>
      <td>14409</td>
      <td>.</td>
      <td>+</td>
      <td>.</td>
      <td>ENSG00000223972</td>
      <td>5</td>
      <td>DDX11L1</td>
      <td>havana</td>
      <td>transcribed_unprocessed_pseudogene</td>
      <td>4</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#output to gtf file and pickle file</span>
<span class="n">gtfpy</span><span class="o">.</span><span class="n">writeGTF</span><span class="p">(</span><span class="n">gtf_all</span><span class="p">,</span><span class="s2">&quot;Homo_sapiens.GRCh38.107.exon.gtf&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">### Convert to `gtf.pkl` File for Fast Loading in Step 2</span>
<span class="n">df_mod_gtf</span><span class="o">=</span><span class="n">gtfpy</span><span class="o">.</span><span class="n">parseGTF</span><span class="p">(</span><span class="n">gtfpy</span><span class="o">.</span><span class="n">readGTF</span><span class="p">(</span><span class="s2">&quot;Homo_sapiens.GRCh38.107.exon.gtf&quot;</span><span class="p">))</span>
<span class="n">df_mod_gtf</span><span class="o">.</span><span class="n">to_pickle</span><span class="p">(</span><span class="s2">&quot;gtf.pkl&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="Tutorials" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="step1_1_preprocess_full_length.html" class="btn btn-neutral float-right" title="Preprocessing Full-Length Single-Cell RNA-Seq for Exon and Junction Read Counts" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright Ding Lab at McGill University.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>