<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Cell Aggregation &mdash; DOLPHIN  documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/copybutton.css" type="text/css" />
      <link rel="stylesheet" href="../_static/nbsphinx-code-cells.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/jquery.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
        <script src="../_static/clipboard.min.js"></script>
        <script src="../_static/copybutton.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Detect Alternative Splicing Events with Outrigger" href="step5_alternative_splicing.html" />
    <link rel="prev" title="Setup and Train the DOLPHIN Model" href="step3_run_DOLPHIN.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            DOLPHIN
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Getting started with DOLPHIN:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../usage.html">Usage</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="step0_generate_exon_gtf.html">Exon-Level Reference GTF File Generation</a></li>
<li class="toctree-l2"><a class="reference internal" href="step1_1_preprocess_full_length.html">Preprocessing Full-Length Single-Cell RNA-Seq for Exon and Junction Read Counts</a></li>
<li class="toctree-l2"><a class="reference internal" href="step1_2_preprocess_10X.html">Preprocessing 10X Single-Cell RNA-Seq for Exon and Junction Read Counts</a></li>
<li class="toctree-l2"><a class="reference internal" href="step2_1_graph_generation.html">Generate Feature Matrix and Adjacency Matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="step2_2_exonGene_matrix_generation.html">Generate Exon Gene h5ad and Select Highly Variable Genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="step2_3_feature_matrix_generation.html">Generate Feature Matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="step2_4_adjacency_matrix_generation.html">Generate Adjacency Matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="step2_5_graph_data_generation.html">Generate Graph Data for MODEl input</a></li>
<li class="toctree-l2"><a class="reference internal" href="step3_run_DOLPHIN.html">Setup and Train the DOLPHIN Model</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Cell Aggregation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Step-1:-Find-Cell-Neighbors-Using-K-Nearest-Neighbors-(KNN)">Step 1: Find Cell Neighbors Using K-Nearest Neighbors (KNN)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Step-2:-Cell-Aggregation---Adding-Junction-Reads-from-Neighboring-Cells">Step 2: Cell Aggregation - Adding Junction Reads from Neighboring Cells</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Get-the-Number-of-Reads-per-BAM-File-for-Library-Size-Normalization">Get the Number of Reads per BAM File for Library Size Normalization</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="step5_alternative_splicing.html">Detect Alternative Splicing Events with Outrigger</a></li>
<li class="toctree-l2"><a class="reference internal" href="step6_alternative_splicing_analysis.html">Alternative Splicing Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="step6_alternative_splicing_analysis.html#Generate-PSI-adata">Generate PSI adata</a></li>
<li class="toctree-l2"><a class="reference internal" href="step6_alternative_splicing_analysis.html#Differential-Alternative-Splicing-Events">Differential Alternative Splicing Events</a></li>
<li class="toctree-l2"><a class="reference internal" href="step7_1_MAST.html">Run MAST to get Exon level p-value</a></li>
<li class="toctree-l2"><a class="reference internal" href="step7_2_EDEG_analysis.html">Exon-Level Differential Gene Analysis Using Feature Matrix</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Examples</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../examples/index.html">Examples</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API Documentations</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../API/index.html">API Documentations</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">DOLPHIN</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">Tutorials</a></li>
      <li class="breadcrumb-item active">Cell Aggregation</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/tutorials/step4_cell_aggregation.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="Cell-Aggregation">
<h1>Cell Aggregation<a class="headerlink" href="#Cell-Aggregation" title="Permalink to this heading"></a></h1>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">from</span> <span class="nn">sklearn.neighbors</span> <span class="kn">import</span> <span class="n">kneighbors_graph</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">shutil</span>
</pre></div>
</div>
</div>
<section id="Step-1:-Find-Cell-Neighbors-Using-K-Nearest-Neighbors-(KNN)">
<h2>Step 1: Find Cell Neighbors Using K-Nearest Neighbors (KNN)<a class="headerlink" href="#Step-1:-Find-Cell-Neighbors-Using-K-Nearest-Neighbors-(KNN)" title="Permalink to this heading"></a></h2>
<p>In this step, we will use the K-Nearest Neighbors (KNN) algorithm to find the nearest neighboring cells for each cell in the dataset.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Load DOLPHIN model results from the h5ad file</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">anndata</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s2">&quot;DOLPHIN_Z.h5ad&quot;</span><span class="p">)</span>

<span class="c1"># Define the number of neighbors (default is 10, including the main cell itself)</span>
<span class="n">n_neighbor</span> <span class="o">=</span> <span class="mi">10</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cell_conn_new</span> <span class="o">=</span> <span class="n">kneighbors_graph</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_z&#39;</span><span class="p">],</span> <span class="mi">10</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;connectivity&#39;</span><span class="p">,</span><span class="n">include_self</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">n_jobs</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span>
<span class="n">cell_dist_new</span> <span class="o">=</span> <span class="n">kneighbors_graph</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_z&#39;</span><span class="p">],</span> <span class="mi">10</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;distance&#39;</span><span class="p">,</span> <span class="n">include_self</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">n_jobs</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#save the neighborhood information for 0.701</span>
<span class="n">main_name</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">combine_name</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">_cell_idx</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;main_sample&quot;</span><span class="p">,</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">_cell_idx</span><span class="p">])</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">_idx</span>  <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nonzero</span><span class="p">(</span><span class="n">cell_conn_new</span><span class="p">[</span><span class="n">_cell_idx</span><span class="p">])[</span><span class="mi">0</span><span class="p">]):</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">_idx</span><span class="p">])</span>
        <span class="n">main_name</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">_cell_idx</span><span class="p">])</span>
        <span class="n">combine_name</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">_idx</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;main_name&quot;</span><span class="p">:</span> <span class="n">main_name</span><span class="p">,</span> <span class="s2">&quot;neighbor&quot;</span><span class="p">:</span><span class="n">combine_name</span><span class="p">})</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;DOLPHIN_aggregation_KNN10.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Step-2:-Cell-Aggregation---Adding-Junction-Reads-from-Neighboring-Cells">
<h2>Step 2: Cell Aggregation - Adding Junction Reads from Neighboring Cells<a class="headerlink" href="#Step-2:-Cell-Aggregation---Adding-Junction-Reads-from-Neighboring-Cells" title="Permalink to this heading"></a></h2>
<p>In this step, we will perform cell aggregation by incorporating confident junction reads from neighboring cells. This process enhances the signal for alternative splicing analysis and helps to resolve potential noise by taking into account the junction read patterns of nearby cells.</p>
</section>
<section id="Get-the-Number-of-Reads-per-BAM-File-for-Library-Size-Normalization">
<h2>Get the Number of Reads per BAM File for Library Size Normalization<a class="headerlink" href="#Get-the-Number-of-Reads-per-BAM-File-for-Library-Size-Normalization" title="Permalink to this heading"></a></h2>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>find<span class="w"> </span>./02_exon_std<span class="w"> </span>-type<span class="w"> </span>f<span class="w"> </span>-name<span class="w"> </span><span class="s2">&quot;*.bam&quot;</span><span class="w"> </span><span class="p">|</span><span class="w"> </span><span class="k">while</span><span class="w"> </span><span class="nb">read</span><span class="w"> </span>file
<span class="k">do</span>
<span class="w">    </span><span class="nb">echo</span><span class="w"> </span><span class="s2">&quot;</span><span class="nv">$file</span><span class="s2">&quot;</span>
<span class="w">    </span><span class="nb">echo</span><span class="w"> </span><span class="s2">&quot;</span><span class="nv">$file</span><span class="s2">&quot;</span><span class="w"> </span>&gt;&gt;<span class="w"> </span>get_single_bam_num.txt
<span class="w">    </span>samtools<span class="w"> </span>flagstat<span class="w"> </span><span class="nv">$file</span><span class="w"> </span>&gt;&gt;<span class="w"> </span>get_single_bam_num.txt
<span class="k">done</span>
</pre></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## Process the number of reads per bam files</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;./get_single_bam_num.txt&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
    <span class="n">lines</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
<span class="n">samples_line</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;/&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">lines</span><span class="p">[</span><span class="mi">0</span><span class="p">::</span><span class="mi">14</span><span class="p">]]</span>
<span class="n">count_line</span> <span class="o">=</span> <span class="n">lines</span><span class="p">[</span><span class="mi">1</span><span class="p">::</span><span class="mi">14</span><span class="p">]</span>
<span class="n">cnt</span> <span class="o">=</span> <span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">count_line</span><span class="p">]</span>
<span class="n">pd_cnt</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;sample&quot;</span><span class="p">:</span><span class="n">samples_line</span><span class="p">,</span> <span class="s2">&quot;num_seqs&quot;</span><span class="p">:</span><span class="n">cnt</span><span class="p">})</span>
<span class="n">pd_cnt</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;single_bam_count.txt&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">metadata</span> <span class="o">=</span> <span class="s2">&quot;your_metaData.csv&quot;</span>
<span class="n">pd_gt</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">metadata</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">sample_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">pd_gt</span><span class="p">[</span><span class="n">pd_gt</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">0</span><span class="p">]])</span>
<span class="n">pd_aggr</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;./DOLPHIN_aggregation_KNN10.csv&quot;</span><span class="p">)</span>
<span class="n">pd_single_size</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;./single_bam_count.txt&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## src_path: path to the original bam and sj files</span>
<span class="n">src_path</span> <span class="o">=</span> <span class="s2">&quot;./02_exon_std/&quot;</span>
<span class="c1">## dist_path: path to the final bam files</span>
<span class="n">dist_path</span> <span class="o">=</span> <span class="s2">&quot;./DOLPHIN_aggregation/&quot;</span>
<br/></pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">target</span> <span class="ow">in</span> <span class="n">sample_list</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">target</span><span class="p">)</span>
    <span class="n">target_size</span> <span class="o">=</span> <span class="n">pd_single_size</span><span class="p">[</span><span class="n">pd_single_size</span><span class="p">[</span><span class="s2">&quot;sample&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">target</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;num_seqs&quot;</span><span class="p">]</span>
    <span class="n">_neighbor</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">pd_aggr</span><span class="p">[</span><span class="n">pd_aggr</span><span class="p">[</span><span class="s2">&quot;main_name&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">target</span><span class="p">][</span><span class="s2">&quot;neighbor&quot;</span><span class="p">])</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dist_path</span><span class="p">,</span> <span class="n">target</span><span class="p">))</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Majority voting: find the frequent junction reads</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="k">for</span> <span class="n">_i</span><span class="p">,</span> <span class="n">_temp_n</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">_neighbor</span><span class="p">):</span>
        <span class="n">_df_junc</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">src_path</span><span class="p">,</span> <span class="n">_temp_n</span><span class="p">,</span> <span class="n">_temp_n</span><span class="o">+</span><span class="s2">&quot;.std.SJ.out.tab&quot;</span><span class="p">),</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">usecols</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">7</span><span class="p">],</span> <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span> <span class="s2">&quot;first_base&quot;</span><span class="p">,</span> <span class="s2">&quot;last_base&quot;</span><span class="p">,</span><span class="s2">&quot;multi_map&quot;</span><span class="o">+</span><span class="n">_temp_n</span><span class="p">])</span>
        <span class="k">if</span> <span class="n">_i</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">df_merge</span> <span class="o">=</span> <span class="n">_df_junc</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">df_merge</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">df_merge</span><span class="p">,</span> <span class="n">_df_junc</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;outer&quot;</span><span class="p">,</span> <span class="n">left_on</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span> <span class="s2">&quot;first_base&quot;</span><span class="p">,</span> <span class="s2">&quot;last_base&quot;</span><span class="p">],</span> <span class="n">right_on</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span> <span class="s2">&quot;first_base&quot;</span><span class="p">,</span> <span class="s2">&quot;last_base&quot;</span><span class="p">])</span>
    <span class="c1">## count the occurence of the neighborhood junctions reads, only keep junction reads which is exist in half of the neighbor cells</span>
    <span class="n">df_merge</span><span class="p">[</span><span class="s2">&quot;nont_na&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">n_neighbor</span> <span class="o">-</span> <span class="n">df_merge</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span> <span class="s2">&quot;first_base&quot;</span><span class="p">,</span> <span class="s2">&quot;last_base&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">isna</span><span class="p">()</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">df_keep_junct</span> <span class="o">=</span> <span class="n">df_merge</span><span class="p">[</span><span class="n">df_merge</span><span class="p">[</span><span class="s2">&quot;nont_na&quot;</span><span class="p">]</span> <span class="o">&gt;=</span><span class="mi">5</span><span class="p">]</span>
    <span class="c1">## save to bed file</span>
    <span class="n">df_keep_junct</span><span class="p">[[</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span> <span class="s2">&quot;first_base&quot;</span><span class="p">,</span> <span class="s2">&quot;last_base&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dist_path</span><span class="p">,</span> <span class="n">target</span><span class="p">,</span> <span class="s2">&quot;keep_junction.bed&quot;</span><span class="p">),</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Bam file batch size normalization</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="k">for</span> <span class="n">_n</span> <span class="ow">in</span> <span class="n">_neighbor</span><span class="p">:</span>
        <span class="n">_n_seq</span> <span class="o">=</span> <span class="n">pd_single_size</span><span class="p">[</span><span class="n">pd_single_size</span><span class="p">[</span><span class="s2">&quot;sample&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">_n</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;num_seqs&quot;</span><span class="p">]</span>
        <span class="n">shutil</span><span class="o">.</span><span class="n">copyfile</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">src_path</span><span class="p">,</span> <span class="n">_n</span><span class="p">,</span> <span class="n">_n</span><span class="o">+</span><span class="s2">&quot;.std.Aligned.sortedByCoord.out.bam&quot;</span><span class="p">),</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dist_path</span><span class="p">,</span> <span class="n">target</span><span class="p">,</span> <span class="n">_n</span><span class="o">+</span><span class="s2">&quot;.bam&quot;</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">_n_seq</span> <span class="o">==</span> <span class="n">target_size</span><span class="p">:</span>
            <span class="n">os</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dist_path</span><span class="p">,</span> <span class="n">target</span><span class="p">,</span> <span class="n">_n</span><span class="o">+</span><span class="s2">&quot;.bam&quot;</span><span class="p">),</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dist_path</span><span class="p">,</span> <span class="n">target</span><span class="p">,</span> <span class="n">_n</span><span class="o">+</span><span class="s2">&quot;.norm.bam&quot;</span><span class="p">))</span>
        <span class="c1">##===== Upsampling:</span>
        <span class="k">elif</span> <span class="n">_n_seq</span> <span class="o">&lt;</span> <span class="n">target_size</span><span class="p">:</span>
            <span class="c1">## random sample some of the sequcen and then add together with original one</span>
            <span class="c1"># concate itself n times, where n is the integer part of target_size/ _n_seq</span>
            <span class="n">_cat_self_n</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">target_size</span><span class="o">/</span> <span class="n">_n_seq</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">_cat_self_n</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">_add_seq_perct</span> <span class="o">=</span> <span class="p">(</span><span class="n">target_size</span> <span class="o">-</span> <span class="n">_n_seq</span><span class="p">)</span><span class="o">/</span><span class="n">_n_seq</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">_add_seq_perct</span> <span class="o">=</span> <span class="p">(</span><span class="n">target_size</span> <span class="o">-</span> <span class="n">_n_seq</span><span class="o">*</span><span class="n">_cat_self_n</span><span class="p">)</span><span class="o">/</span><span class="n">_n_seq</span>
            <span class="c1">## sample the reset seq reads</span>
            <span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;samtools view -b -s </span><span class="si">{</span><span class="n">_add_seq_perct</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dist_path</span><span class="p">,</span><span class="w"> </span><span class="n">target</span><span class="p">,</span><span class="w"> </span><span class="n">_n</span><span class="o">+</span><span class="s1">&#39;.bam&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2"> &gt; </span><span class="si">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dist_path</span><span class="p">,</span><span class="w"> </span><span class="n">target</span><span class="p">,</span><span class="w"> </span><span class="n">_n</span><span class="o">+</span><span class="s1">&#39;.sample.bam&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="c1">## concatenate all</span>
            <span class="n">combine_name</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
            <span class="n">current_name</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dist_path</span><span class="p">,</span> <span class="n">target</span><span class="p">,</span> <span class="n">_n</span><span class="o">+</span><span class="s1">&#39;.bam&#39;</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">_cat_self_n</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">combine_name</span> <span class="o">=</span> <span class="n">current_name</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">combine_name</span> <span class="o">=</span> <span class="n">combine_name</span> <span class="o">+</span> <span class="s2">&quot; &quot;</span> <span class="o">+</span> <span class="n">current_name</span>
            <span class="n">combine_name</span> <span class="o">=</span> <span class="n">combine_name</span> <span class="o">+</span> <span class="s2">&quot; &quot;</span> <span class="o">+</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dist_path</span><span class="p">,</span> <span class="n">target</span><span class="p">,</span> <span class="n">_n</span><span class="o">+</span><span class="s1">&#39;.sample.bam&#39;</span><span class="p">)</span>
            <span class="n">result_name</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dist_path</span><span class="p">,</span> <span class="n">target</span><span class="p">,</span> <span class="n">_n</span><span class="o">+</span><span class="s2">&quot;.norm.bam&quot;</span><span class="p">)</span>
            <span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;samtools merge </span><span class="si">{</span><span class="n">result_name</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">combine_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dist_path</span><span class="p">,</span> <span class="n">target</span><span class="p">,</span> <span class="n">_n</span><span class="o">+</span><span class="s2">&quot;.sample.bam&quot;</span><span class="p">))</span>
            <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dist_path</span><span class="p">,</span> <span class="n">target</span><span class="p">,</span> <span class="n">_n</span><span class="o">+</span><span class="s2">&quot;.bam&quot;</span><span class="p">))</span>
        <span class="c1">##===== Downsampling:</span>
        <span class="k">if</span> <span class="n">_n_seq</span> <span class="o">&gt;</span> <span class="n">target_size</span><span class="p">:</span>
            <span class="n">_keep_seq_perct</span> <span class="o">=</span> <span class="n">target_size</span><span class="o">/</span><span class="n">_n_seq</span>
            <span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;samtools view -b -s </span><span class="si">{</span><span class="n">_keep_seq_perct</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dist_path</span><span class="p">,</span><span class="w"> </span><span class="n">target</span><span class="p">,</span><span class="w"> </span><span class="n">_n</span><span class="o">+</span><span class="s1">&#39;.bam&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2"> &gt; </span><span class="si">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dist_path</span><span class="p">,</span><span class="w"> </span><span class="n">target</span><span class="p">,</span><span class="w"> </span><span class="n">_n</span><span class="o">+</span><span class="s1">&#39;.norm.bam&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dist_path</span><span class="p">,</span> <span class="n">target</span><span class="p">,</span> <span class="n">_n</span><span class="o">+</span><span class="s2">&quot;.bam&quot;</span><span class="p">))</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        Bam file split to junction readsfile</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="k">if</span> <span class="n">_n</span> <span class="o">!=</span> <span class="n">target</span><span class="p">:</span>
            <span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;samtools view -h </span><span class="si">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dist_path</span><span class="p">,</span><span class="w"> </span><span class="n">target</span><span class="p">,</span><span class="w"> </span><span class="n">_n</span><span class="o">+</span><span class="s1">&#39;.norm.bam&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2"> | awk &#39;$0 ~ /^@/ || $6 ~ /N/&#39; | samtools view -b &gt; </span><span class="si">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dist_path</span><span class="p">,</span><span class="w"> </span><span class="n">target</span><span class="p">,</span><span class="w"> </span><span class="n">_n</span><span class="o">+</span><span class="s1">&#39;.junction.norm.bam&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;samtools index </span><span class="si">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dist_path</span><span class="p">,</span><span class="w"> </span><span class="n">target</span><span class="p">,</span><span class="w"> </span><span class="n">_n</span><span class="o">+</span><span class="s1">&#39;.junction.norm.bam&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="w">            </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">            Filter to only keep frequent junctions</span>
<span class="sd">            &#39;&#39;&#39;</span>
            <span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;samtools view -h -L </span><span class="si">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dist_path</span><span class="p">,</span><span class="w"> </span><span class="n">target</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;keep_junction.bed&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dist_path</span><span class="p">,</span><span class="w"> </span><span class="n">target</span><span class="p">,</span><span class="w"> </span><span class="n">_n</span><span class="o">+</span><span class="s1">&#39;.junction.norm.bam&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2"> &gt; </span><span class="si">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dist_path</span><span class="p">,</span><span class="w"> </span><span class="n">target</span><span class="p">,</span><span class="w"> </span><span class="n">_n</span><span class="o">+</span><span class="s1">&#39;.mj.junction.norm.bam&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Final concate all normalized bam files</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="c1"># Final concate all normalized fastq files</span>
    <span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;samtools merge </span><span class="si">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dist_path</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;final_bam&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">target</span><span class="o">+</span><span class="s1">&#39;.final.bam&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dist_path</span><span class="p">,</span><span class="w"> </span><span class="n">target</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;*.mj.junction.norm.bam&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">  </span><span class="si">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dist_path</span><span class="p">,</span><span class="w"> </span><span class="n">target</span><span class="p">,</span><span class="w"> </span><span class="n">target</span><span class="o">+</span><span class="s1">&#39;.norm.bam&#39;</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">shutil</span><span class="o">.</span><span class="n">rmtree</span><span class="p">((</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dist_path</span><span class="p">,</span> <span class="n">target</span><span class="p">)))</span>
</pre></div>
</div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="step3_run_DOLPHIN.html" class="btn btn-neutral float-left" title="Setup and Train the DOLPHIN Model" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="step5_alternative_splicing.html" class="btn btn-neutral float-right" title="Detect Alternative Splicing Events with Outrigger" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright Ding Lab at McGill University.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>